---
title: "Clustering on All T Cell CROP-seq Data"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(Seurat)
# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
wkdir <- "/project2/xinhe/yifan/Factor_analysis/Stimulated_T_Cells/"
```

# Data Settings

Source:   
[Genome-wide CRISPR Screens in Primary Human T Cells Reveal Key Regulators of Immune Function](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6689405/), GEO accession: GSE119450.

**The following analysis follows a similar approach as used in this reference paper to generate Fig 4.**

Perturbations:    
CRISPR knock-out of 20 genes (2 gRNAs per gene) + 8 non-targeting gRNAs. Guide conditions were defined on the target gene level; target genes were either found to regulate T cell responses in the genome-wide screens, or known checkpoint genes.    
Guide RNAs were introduced into T cells through a novel procedure called sgRNA lentiviral infection with Cas9 protein electroporation (SLICE).

Cells:    
Primary human CD8+ T cells from two healthy donors, with and without T cell receptor (TCR) stimulation.   
All cells were pooled together into 1 analysis.

# scRNA-seq data

## Preprocessing

There are 24955 cells with a single type of gRNA readout using the quality control criteria of % mitochondria gene expression < 10 and number of unique genes > 500.

```{r load seurat object}
combined_obj <-
  readRDS(paste0(wkdir, "processed_data/seurat_obj.all_T_cells_merged.uncorrected.rds"))
```

```{r QC, fig.height=5}
VlnPlot(combined_obj, features = c("nFeature_RNA", "nCount_RNA", "percent_mt"), pt.size = 0.1)
```

Gene expressions were normalized as log(count per $10^4$ + 1), and then scaled.

**Library size and mitochondria percentage were corrected for, while batch effect was not.**

Top 1000 most variable genes were used to perform PCA, and the first 30 PCs were used to perform UMAP.

## UMAP representation

Donors and stimulation status:

```{r umap_batch, fig.width=7, fig.height=5}
DimPlot(combined_obj, reduction = "umap", group.by = "orig.ident",
        cols = c("red3", "steelblue4", "forestgreen", "orchid4")) +
  labs(title = "All T Cells")
```

Clustering of cells:
```{r umap_cell_clusters, fig.width=7, fig.height=5}
combined_obj <- FindClusters(combined_obj, resolution = 0.45, verbose = F)
cluster_plot_df <- DimPlot(combined_obj, reduction = "umap", cols = "Set3")$data
levels(cluster_plot_df$ident) <- c(8, 1, 6, 2, 5, 4, 3, 7)
cluster_plot_df$ident <- factor(cluster_plot_df$ident, levels = 1:8)
label_df <- cluster_plot_df %>% group_by(ident) %>%
  summarise(xpos = mean(UMAP_1), ypos = mean(UMAP_2))
ggplot(cluster_plot_df, aes(x = UMAP_1, y = UMAP_2, color = ident)) +
  geom_point(size = 0.5) +
  scale_color_brewer(palette = "Set3") +
  geom_text(data = label_df, aes(x = xpos, y = ypos, label = ident), color = "black") +
  labs(x = "UMAP 1", y = "UMAP 2",
       title = "Clustering of All T Cells", color = "Cluster") +
  theme(panel.grid = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 2)))
combined_meta <- combined_obj@meta.data
combined_meta$seurat_clusters <- cluster_plot_df$ident
```

Expression patterns of marker genes:    
activation state (IL7R, CCR7), cell cycle (MKI67), and effector function (GZMB) in the reference paper.
```{r umap_marker_genes, fig.height=8, fig.width=12}
feature.names <- data.frame(fread(paste0(wkdir, "GSE119450_RAW/D1S/genes.tsv"),
                                  header = FALSE), stringsAsFactors = FALSE)
norm.rna <- combined_obj@assays$RNA@data
marker_genes <- c("IL7R", "CCR7" , "MKI67", "GZMB")
for (m in marker_genes){
  gene_id <- feature.names$V1[feature.names$V2 == m]
  cluster_plot_df[[m]] <- as.numeric(norm.rna[gene_id, ])
  cluster_plot_df[[m]][cluster_plot_df[[m]] > 3] <- 3
}
plot_list <- list()
for (m in marker_genes){
  plot_list[[m]] <- ggplot(cluster_plot_df,
                           aes_string(x = "UMAP_1", y = "UMAP_2", color = m)) +
    geom_point(size = 0.3) +
    scale_color_gradientn(colors = c("#91bfdb", "#ffffbf", "#fc8d59")) +
    labs(x = "UMAP 1", y = "UMAP 2", title = m, color = "log(normalized\nexpression)") +
    theme(panel.grid = element_blank())
}
do.call(grid.arrange, plot_list)
```

Cluster 1, 2 and 3: resting state

Cluste 5 and 6: proliferation

# CRISPR Perturbations

```{r num_gRNA_cells, fig.width=8, fig.height=5}
num_cells_per_guide <- colSums(combined_meta[, 4:24])
num_cells.df <- data.frame(locus = names(num_cells_per_guide),
                           count = num_cells_per_guide)
ggplot(data = num_cells.df, aes(x=locus, y=count)) +
  geom_bar(stat = "identity", width = 0.6) +
  labs(title = "Cell count per gene KO target", x = "KO Target", y = "# of cells") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12))
```

## Over/Under-Representation of gRNA Targets in Clusters

```{r chisq_test}
targets <- colnames(combined_meta)[4:24]
chisq_stdres <- matrix(nrow = length(levels(combined_meta$seurat_clusters)),
                       ncol = length(targets))
colnames(chisq_stdres) <- targets
rownames(chisq_stdres) <- paste0("Cluster ", 1:nrow(chisq_stdres))
chisq_pval <- chisq_stdres
for (m in targets){
  test <- chisq.test(table(factor(combined_meta[[m]]), combined_meta$seurat_clusters))
  chisq_stdres[, m] <- test$stdres[2, ]
  for (k in 1:nrow(chisq_stdres)){
    chisq_pval[k, m] <- chisq.test(table(factor(combined_meta[[m]]),
                                         combined_meta$seurat_clusters == k))$p.value
  }
}
# saveRDS(list(chisq_stdres = chisq_stdres, chisq_pval = chisq_pval),
#         paste0(wkdir, "processed_data/chisq_test_stats.all_T_cells_merged.rds"))
```

```{r chisq_test_pval_heatmap, fig.height=5.5, fig.width=7}
Heatmap(t(-log10(chisq_pval) * sign(chisq_stdres)),
        name = "chi-square polarized\n-log10(p value)",
        cluster_rows = F, cluster_columns = F)
```

Pairs of associations that passed chi-square test FDR < 0.05:

```{r}
chisq_pval_adj <- p.adjust(unlist(chisq_pval), method = "fdr")
cat(sum(chisq_pval_adj < 0.05))
cat("\n\n")
```

```{r}
chisq_pval_adj <- matrix(chisq_pval_adj, nrow = nrow(chisq_pval))
rownames(chisq_pval_adj) <- rownames(chisq_pval)
colnames(chisq_pval_adj) <- colnames(chisq_pval)
signif_guides <- colnames(chisq_pval_adj)[colSums(chisq_pval_adj < 0.05) > 0]
cat("Perturbations with significant enrichment or depletion in clusters:\n")
cat(paste(signif_guides, collapse = ", "))
```

```{r fig.width=6.5, fig.height=4, eval=F, include=F}
make_pointwise_cdf <- function(pval_vec, bonf_cutoff = 0.05){
  pval_thres <- bonf_cutoff / length(pval_vec)
  neg_pval_vec <- -log10(pval_vec)
  xvec <- neg_pval_vec[order(neg_pval_vec)] # seq(0, 10, 0.5)
  cdf_df <- data.frame(neg_pval = xvec,
                       cdf = sapply(xvec, function(x){mean(neg_pval_vec < x)}))
  cdf_df <- cdf_df %>% mutate(significant = neg_pval > -log10(pval_thres)) %>%
    mutate(significant = factor(significant, levels = c(TRUE, FALSE)))
  return(cdf_df)
}
cdf_df <- make_pointwise_cdf(unlist(chisq_pval))
ggplot(cdf_df, aes(x = neg_pval, y = cdf, color = significant)) +
  geom_point(size = 0.8) +
  scale_color_manual(values = c("red", "black")) +
  labs(x = expression(-log[10]*"(Chi-square p value)"),
       y = "CDF",
       color = "Bonferroni < 0.05") +
  guides(color = guide_legend(override.aes = list(size = 2)))
```

```{r chisq_test_heatmap}
stdres_colormap <- circlize::colorRamp2(breaks = c(-3, 0, 3),
                                        colors = c("blue", "white", "red"))
Heatmap(t(chisq_stdres),
        col = stdres_colormap,
        name = "chi-square \nstandardized residual",
        cluster_rows = T, cluster_columns = T)
```

Selected clusters and gRNA targets:

```{r chisq_test_barplot_selected, fig.width=10, fig.height=7}
cluster_indx <- paste0("Cluster ", c(1, 2, 5, 6))
# marker_indx <- c("ARID1A", "CBLB", "CD5", "CDKN1B", "SOCS1", "TCEB2",
#                  "RASA2", "LCP2", "CD3D", "NonTarget")
marker_indx <- c("ARID1A", "CBLB", "CD5", "TCEB2",
                 "RASA2", "LCP2", "CD3D", "NonTarget")
marker_indx <- marker_indx[length(marker_indx):1]
selected_stdres_df <- as.data.frame(t(chisq_stdres[cluster_indx, marker_indx])) %>%
  mutate(target = factor(marker_indx, levels = marker_indx))
selected_stdres_df <- reshape2::melt(selected_stdres_df, id.var = "target",
                                     variable.name = "cluster", value.name = "stdres")

selected_pval_df <- as.data.frame(t(chisq_pval[cluster_indx, marker_indx])) %>%
  mutate(target = factor(marker_indx, levels = marker_indx))
selected_pval_df <- reshape2::melt(selected_pval_df, id.var = "target",
                                   variable.name = "cluster", value.name = "pval")
combined_selected_df <- inner_join(selected_stdres_df, selected_pval_df,
                                   by = c("target", "cluster")) %>%
  mutate(neg_logp = -log10(pval)) %>%
  mutate(sign_logp = ifelse(stdres > 0, neg_logp, -neg_logp))
pval_range <- range(combined_selected_df$sign_logp)
ggplot(combined_selected_df,
       aes(y = target, x = stdres, fill = sign_logp)) +
  facet_wrap(~ cluster) +
  geom_col() +
  # scale_fill_gradient2(low = scales::muted("blue"), high = scales::muted("red")) +
  scale_fill_gradientn(colors = c(scales::muted("blue"), scales::muted("blue"), "white",
                                  scales::muted("red"), scales::muted("red")),
                       values = scales::rescale(c(pval_range[1], -4, 0, 4, pval_range[2])),
                       breaks = seq(-6, 6, 3)) +
  labs(x = "Chi-square test stardardized residual",
       fill = "Polarized \n-log10(p value)") +
  theme(axis.title.y = element_blank(),
        strip.text = element_text(size = 14))
```

# Stimulated cells only

```{r load_stim_seurat_preprocessed}
# stim_obj <- readRDS(paste0(wkdir,
#                            "processed_data/seurat_obj.stimulated_merged.dev_top6k.corrected_3.rds"))
# stim_obj <- FindVariableFeatures(stim_obj)
# stim_obj <- ScaleData(stim_obj,
#                       vars.to.regress = c("nCount_RNA", "percent_mt"))
# stim_obj <- RunPCA(stim_obj, verbose = F)
# stim_obj <- FindNeighbors(stim_obj, reduction = "pca")
# stim_obj <- RunUMAP(stim_obj, reduction = "pca", dims = 1:20)
# DimPlot(stim_obj, reduction = "umap", group.by = "orig.ident",
#         cols = c("steelblue4", "orchid4"))
# saveRDS(stim_obj,
#         paste0(wkdir, "processed_data/seurat_obj.stimulated_merged.top2k_var.corrected_2.rds"))
stim_obj <- readRDS(paste0(wkdir, "processed_data/seurat_obj.stimulated_merged.top2k_var.corrected_2.rds"))
DimPlot(stim_obj, reduction = "umap", group.by = "orig.ident",
        cols = c("steelblue4", "orchid4")) +
  labs(title = "Stimulated Cells")
```

## Clustering of cells
```{r stim_cell_clusters, fig.width=7, fig.height=5}
stim_obj <- FindClusters(stim_obj, resolution = 0.3, verbose = F)
cluster_plot_df <- DimPlot(stim_obj, reduction = "umap", cols = "Set3")$data
levels(cluster_plot_df$ident) <- c(5, 1, 2, 4, 3)
cluster_plot_df$ident <- factor(cluster_plot_df$ident, levels = 1:5)
label_df <- cluster_plot_df %>% group_by(ident) %>%
  summarise(xpos = mean(UMAP_1), ypos = mean(UMAP_2))
ggplot(cluster_plot_df, aes(x = UMAP_1, y = UMAP_2, color = ident)) +
  geom_point(size = 0.5) +
  scale_color_brewer(palette = "Set3") +
  # geom_text(data = label_df, aes(x = xpos, y = ypos, label = ident), color = "black") +
  labs(x = "UMAP 1", y = "UMAP 2",
       title = "Clustering of Stimulated T Cells", color = "Cluster") +
  theme(panel.grid = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 2)))
```

Expression patterns of marker genes:    
activation state (IL7R, CCR7), cell cycle (MKI67), and effector function (GZMB) in the reference paper.
```{r fig.height=8, fig.width=12}
feature.names <- data.frame(fread(paste0(wkdir, "GSE119450_RAW/D1S/genes.tsv"),
                                  header = FALSE), stringsAsFactors = FALSE)
norm.rna <- stim_obj@assays$RNA@data
marker_genes <- c("IL7R", "CCR7" , "MKI67", "GZMB")
for (m in marker_genes){
  gene_id <- feature.names$V1[feature.names$V2 == m]
  cluster_plot_df[[m]] <- as.numeric(norm.rna[gene_id, ])
  cluster_plot_df[[m]][cluster_plot_df[[m]] > 3] <- 3
}
plot_list <- list()
for (m in marker_genes){
  plot_list[[m]] <- ggplot(cluster_plot_df,
                           aes_string(x = "UMAP_1", y = "UMAP_2", color = m)) +
    geom_point(size = 0.3) +
    scale_color_gradientn(colors = c("#91bfdb", "#ffffbf", "#fc8d59")) +
    labs(x = "UMAP 1", y = "UMAP 2", title = m, color = "log(normalized\nexpression)") +
    theme(panel.grid = element_blank())
}
do.call(grid.arrange, plot_list)
```

## Over/Under-Representation of gRNA Targets in Clusters

```{r}
sub_meta <- stim_obj@meta.data
sub_meta$seurat_clusters <- cluster_plot_df$ident
targets <- colnames(sub_meta)[4:24]
chisq_stdres <- matrix(nrow = length(levels(sub_meta$seurat_clusters)),
                       ncol = length(targets))
colnames(chisq_stdres) <- targets
rownames(chisq_stdres) <- paste0("Cluster ", 1:nrow(chisq_stdres))
chisq_pval <- chisq_stdres
for (m in targets){
  test <- chisq.test(table(factor(sub_meta[[m]]), sub_meta$seurat_clusters))
  chisq_stdres[, m] <- test$stdres[2, ]
  for (k in 1:nrow(chisq_stdres)){
    chisq_pval[k, m] <- chisq.test(table(factor(sub_meta[[m]]),
                                         sub_meta$seurat_clusters == k))$p.value
  }
}
```

```{r fig.height=5.5, fig.width=6}
Heatmap(t(-log10(chisq_pval) * sign(chisq_stdres)),
        name = "chi-square polarized\n-log10(p value)",
        cluster_rows = F, cluster_columns = F)
```

Pairs of associations that passed chi-square test FDR < 0.05:

```{r}
chisq_pval_adj <- p.adjust(unlist(chisq_pval), method = "fdr")
cat(sum(chisq_pval_adj < 0.05))
cat("\n\n")
```

```{r}
chisq_pval_adj <- matrix(chisq_pval_adj, nrow = nrow(chisq_pval))
rownames(chisq_pval_adj) <- rownames(chisq_pval)
colnames(chisq_pval_adj) <- colnames(chisq_pval)
signif_guides <- colnames(chisq_pval_adj)[colSums(chisq_pval_adj < 0.05) > 0]
cat("Perturbations with significant enrichment or depletion in clusters:\n")
cat(paste(signif_guides, collapse = ", "))
```

```{r eval=F, include=F}
## Percentage of cells w/ guide in each cluster
colMeans((sub_meta %>% dplyr::filter(seurat_clusters == "4"))[, 4:24])
```
