---
title: "Guided Factor Analysis on LUHMES CROP-seq Data"
subtitle: "-- NTC included, unique UMI count corrected"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(clusterProfiler)

# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
print_signif_tb <- function(signif_num, caption_text){
  if (length(signif_num) < 15){
    signif_num <- c(signif_num, rep(NA, 15 - length(signif_num)))
  }
  signif_num <- t(data.frame(KO = names(signif_num),
                             Num_genes = signif_num,
                             row.names = NULL))
  print(knitr::kable(rbind(signif_num[, 1:5], signif_num[, 6:10], signif_num[, 11:15]),
                     caption = caption_text) %>%
    kable_styling() %>% scroll_box(width = '100%'))
}
print_lfsr_signif_tb <- function(lfsr_mat, lfsr_cutoff, caption_text){
  signif_num <- colSums(lfsr_mat < lfsr_cutoff)
  print_signif_tb(signif_num, caption_text)
}

wkdir <- "/project2/xinhe/yifan/Factor_analysis/LUHMES/"
gsfa_folder <- "gsfa_output_detect_01/merged_new/"
guide <- "All"
pip_col_fun <- circlize::colorRamp2(breaks = c(0, 0.5, 1),
                                    colors = c("black", "purple", "gold"))
```

# Single cell expression data

Source:   
[High-throughput single-cell functional elucidation of neurodevelopmental disease-associated genes reveals convergent mechanisms altering neuronal differentiation](https://genome.cshlp.org/content/30/9/1317.full), GEO accession: GSE142078.

Perturbations:    
CRISPR knock-down of 14 autism spectrum disorder (ASD)â€“associated genes (3 gRNAs per gene) + 5 non-targeting gRNAs.

Cells:    
Lund human mesencephalic (LUHMES) neural progenitor cell line.   
Cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout. Quality control resulted in 8708 cells.

Genes:    
**Only genes detected in > 10% of cells were kept, resulted in 6213 genes.**

Normalization:    
Seurat "LogNormalize": log(count per 10K + 1).   
**Batch effect, unique UMI count, library size, and mitochondria percentage were all corrected for. The corrected and scaled expression data were used as input for subsequent factor analysis.**

```{r load meta data, fig.width=9}
metadata <- readRDS(paste0(wkdir, "processed_data/merged_metadata.rds"))
all_KOs.G_mat <- metadata[, 4:18]
KO_names <- colnames(all_KOs.G_mat)
```

# Guided results

Here, our "guide", $G$ matrix, consists of 15 types (14 genes + NTC) of KO conditions across cells.

```{r}
if (guide == "no_NTC"){
  KO_names <- KO_names[KO_names != "Nontargeting"]
  all_KOs.G_mat <- all_KOs.G_mat[, KO_names]
}
nmarkers <- ncol(all_KOs.G_mat)
```

In each case, Gibbs sampling was conducted for 2000 iterations, and the posterior mean estimates were averaged over the last 500 iterations.

## SVD Initialization
```{r svd_tb}
All_KOs.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.restart.PM.rds"))
stopifnot(nrow(All_KOs.gibbs_PM$beta_pm) == nmarkers + 1)
All_KOs.gibbs_res_tb <- make_gibbs_res_tb(All_KOs.gibbs_PM, all_KOs.G_mat, compute_pve = F)
```

### Local False Sign Rate (LFSR)

For a given GSFA inference result, we can estimate the effect a certain KO condition $m$ has on the expression of gene $j$ by computing the LFSR of $\beta_{m\cdot} \cdot W_{j\cdot}$.

```{r}
lfsr_mat <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.restart.lfsr_mat.rds"))
print_lfsr_signif_tb(lfsr_mat, 0.05,
                     paste0("GSFA, # of genes that passed LFSR cutoff of 0.05:"))
```

### Factor ~ KO Beta PIP
```{r fig.width=8}
beta_pip_matrix <- t(All_KOs.gibbs_PM$Gamma_pm[-(nmarkers + 1), ])
rownames(beta_pip_matrix) <- 1:nrow(beta_pip_matrix)
colnames(beta_pip_matrix) <- colnames(all_KOs.G_mat)
Heatmap(beta_pip_matrix, name = "Beta PIP", col = pip_col_fun,
        row_title = "Factors",  column_title = "KO Perturbations",
        cluster_rows = F, cluster_columns = F)
```

### Factor ~ KO Posterior Association
```{r fig.width=10}
heatmap_matrix <- All_KOs.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = All_KOs.gibbs_res_tb$pi)
guided.heatmap_matrix <- heatmap_matrix
gsfa_KO_names <- apply(guided.heatmap_matrix, 1,
                       function(x){ paste(colnames(all_KOs.G_mat)[x < 1e-4], collapse = ",") })
gsfa_KO_names <- paste(1:nrow(guided.heatmap_matrix), gsfa_KO_names, sep = " - ")
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA All KOs (K=20, SVD)")
```

### Beta PIP vs P-Value
```{r fig.width=12, fig.height=5}
p1 <- qplot(unlist(beta_pip_matrix), -log10(unlist(heatmap_matrix)),
            xlab = "Beta PIP", ylab = "-log10(Association p-value)",
            main = "20 factors x 15 conditions")
p2 <- qplot(unlist(beta_pip_matrix), -log10(unlist(heatmap_matrix)),
            ylim = c(0, 10),
            xlab = "Beta PIP", ylab = "-log10(Association p-value)",
            main = "Y-axis truncated at 10")
grid.arrange(p1, p2, nrow = 1)
```

### Correlation btw Factors

```{r eval=FALSE, include=FALSE}
plot_pairwise.corr_heatmap(input_mat_1 = All_KOs.gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Pairwise Correlation in Sample Loadings - Z (GSFA)")
```

```{r}
plot_pairwise.corr_heatmap(input_mat_1 = (All_KOs.gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Binarized Gene Loadings (F.pm > 0.95)")
```

## Alternative Initializations and Chain Convergence

Other than initializing GSFA using truncated SVD, we also ran 2 chains with random initialization.

```{r}
rand_1.gibbs_PM <- 
  readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.rand_01.restart.PM.rds"))
rand_2.gibbs_PM <-
  readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.rand_02.restart.PM.rds"))
```

### Association results for 2 randomly initialized GSFA runs

```{r rand_01, fig.width=10, fig.height=5}
rand_1.gibbs_res_tb <- make_gibbs_res_tb(rand_1.gibbs_PM, all_KOs.G_mat, compute_pve = F)
# rand_1.gibbs_res_tb <- rand_1.gibbs_res_tb %>% arrange(-sum_var)
heatmap_matrix <- rand_1.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- rand_1.gibbs_res_tb$index
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
# plot_pval_heatmap(heatmap_matrix,
#                   factor_annot = rand_1.gibbs_res_tb$pi)
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA All KOs (K=20, Random 1)")
```

```{r rand_02, fig.width=10, fig.height=5}
rand_2.gibbs_res_tb <- make_gibbs_res_tb(rand_2.gibbs_PM, all_KOs.G_mat, compute_pve = F)
# rand_2.gibbs_res_tb <- rand_2.gibbs_res_tb %>% arrange(-sum_var)
heatmap_matrix <- rand_2.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- rand_2.gibbs_res_tb$index
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
# plot_pval_heatmap(heatmap_matrix,
#                   factor_annot = rand_2.gibbs_res_tb$pi)
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA All KOs (K=20, Random 2)")
```

### Chain mixing and convergence evaluation

```{r show_R_hat_summary}
converg_mat <-
  readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.beta_W_converg_mat.1k-2k.rds"))
knitr::kable(signif(t(apply(converg_mat, 2, summary)), digits = 3)) %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Difference btw methods in factor estimation

For a pair of $Z$ estimations from 2 inference methods, $Z_1, Z_2$, we quantify the pairwise estimation difference as $||Z_1Z_1^T - Z_2Z_2^T||_F/N$, where $||\cdot||$ is the Frobenius norm of a matrix, and $N$ is the number of rows (samples) in $Z$.

```{r}
cat("Guided SVD vs Rand_01:",
    ZZT_mean_diff(All_KOs.gibbs_PM$Z_pm, rand_1.gibbs_PM$Z_pm) %>% signif(digits = 3))
cat("\n\n")
cat("Guided SVD vs Rand_02:",
    ZZT_mean_diff(All_KOs.gibbs_PM$Z_pm, rand_2.gibbs_PM$Z_pm) %>% signif(digits = 3))
cat("\n\n")
cat("Rand_01 vs Rand_02:",
    ZZT_mean_diff(rand_1.gibbs_PM$Z_pm, rand_2.gibbs_PM$Z_pm) %>% signif(digits = 3))
cat("\n\n")
# cat("Guided SVD vs Unguided:",
#     ZZT_mean_diff(All_KOs.gibbs_PM$Z_pm, G0.gibbs_PM$Z_pm) %>% signif(digits = 3))
# cat("\n\n")
```

# Unguided Result

In contrast to the guided version, here we provide no guidance ($G = \vec{0}$) to our sparse factor analysis model to serve as a comparison. 
```{r G0_tb}
G0.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, "G0.gibbs_obj_k20.svd.PM.rds"))
G0.gibbs_res_tb <- make_gibbs_res_tb(G0.gibbs_PM, all_KOs.G_mat, compute_pve = F)
```

## Factor ~ KO Beta PIP
```{r fig.width=8, fig.height=1}
beta_pip_matrix <- t(G0.gibbs_PM$Gamma_pm[-2, ])
colnames(beta_pip_matrix) <- 1:ncol(beta_pip_matrix)
rownames(beta_pip_matrix) <- "G0"
Heatmap(beta_pip_matrix,
        name = "Beta PIP", col = pip_col_fun,column_title = "Factors",
        cluster_rows = F, cluster_columns = F,
        heatmap_legend_param = list(legend_direction = "horizontal"))
```

## Factor ~ KO Posterior Association
```{r fig.width=10}
heatmap_matrix <- G0.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = G0.gibbs_res_tb$pi)
G0.heatmap_matrix <- heatmap_matrix
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA Unguide All KOs (K=20, SVD)")
```

## Unguided vs Guided

```{r fig.width=8}
paired_pval_ranked_scatterplot(pval_vec_1 = unlist(guided.heatmap_matrix),
                               pval_vec_2 = unlist(G0.heatmap_matrix),
                               name_1 = "Guided", name_2 = "Unguided")
```

# FLASH

```{r flash_tb}
flash_obj <- readRDS(paste0(wkdir, "flash_output/merged_scaled_detect_01.corrected_new.flashier_obj_k20.rds"))
flash_res_tb <- make_flash_res_tb(flash_obj, all_KOs.G_mat)
```

## Factor ~ KO Posterior Association

```{r FLASH pval heatmap, fig.width=10}
heatmap_matrix <- flash_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = flash_res_tb$pi1)
flash_heatmap <- heatmap_matrix
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "FLASH All KOs (K=20, SVD)")
```

## FLASH vs GSFA

```{r fig.width=8}
paired_pval_ranked_scatterplot(pval_vec_1 = unlist(guided.heatmap_matrix),
                               pval_vec_2 = unlist(flash_heatmap),
                               name_1 = "Guided", name_2 = "FLASH")
```

We matched the factors obtained from GFSA with those from FLASH, and conducted pairwise comparisons for both factors and gene loadings.

**In the following heatmaps, factors are labeled by the KO condition(s) that they have an association p-value < 1e-4 with.**

```{r}
flash_KO_names <- apply(flash_heatmap, 1,
                        function(x){ paste(colnames(all_KOs.G_mat)[x < 1e-4], collapse = ",") })
flash_KO_names <- paste(1:nrow(flash_heatmap), flash_KO_names, sep = " - ")
# flash_matching_order <- apply(corr_mat, 1, function(x){ which.max(abs(x)) })
flash_matching_order <- c(1, 2, 3, 4, 5, 8, 6, 7, 9, 15, 14, 12, 15, 11, 10, 13, 18, 19, 16, 13)
```

```{r fig.width=9, fig.height=7}
Z_pm.flash <- flash_obj$loadings.pm[[1]][, flash_matching_order]
colnames(Z_pm.flash) <- flash_KO_names[flash_matching_order]
Z_pm.gsfa <- All_KOs.gibbs_PM$Z_pm
colnames(Z_pm.gsfa) <- gsfa_KO_names
plot_pairwise.corr_heatmap(input_mat_1 = Z_pm.gsfa,
                           input_mat_2 = Z_pm.flash,
                           name_1 = "GSFA Factors (Z_pm)", name_2 = "FLASH Factors")
```

```{r fig.width=9, fig.height=7}
F_pm.flash <- flash_obj$loadings.lfsr[[2]][, flash_matching_order]
colnames(F_pm.flash) <- flash_KO_names[flash_matching_order]
F_pm.gsfa <- All_KOs.gibbs_PM$F_pm
colnames(F_pm.gsfa) <- gsfa_KO_names
plot_pairwise.corr_heatmap(input_mat_1 = F_pm.gsfa > 0.95,
                           input_mat_2 = F_pm.flash < 0.05,
                           name_1 = "GSFA Binarized Gene Loadings (F_pm > 0.95)",
                           name_2 = "FLASH Binarized Gene Loadings (LFSR < 0.05)",
                           corr_type = "jaccard")
```

# Gene Set Enrichment Analysis

Target genes: Genes w/ non-zero loadings in each factor (PIP cutoff at 0.95);    
Backgroud genes: all 6213 genes used in factor analysis;    
Statistical test: hypergeometric test (over-representation test);    
Only GO terms/pathways that satisfy fold change $\geq$ 2 and test FDR $<$ 0.05 are shown below.

## GO Slim Over-Representation Analysis

Gene sets: [Gene ontology](http://www.geneontology.org/) "Biological Process" (non-redundant).

```{r}
gsfa.Web_ORA_go_bp <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.GO_bp.cutoff_0.95.rds"))
signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_go_bp, enrich_type = "GO terms")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Factors of interest

10: (nerve development, cell fate commitment);   
11: chromatin remodeling, DNA recombination;   
**13**: regulation of metal ion transport, axon development, ...;   
**15**: neuron migration, response to metal ion, CNS neuron development, ...;   
**18**: CNS neuron differentiation, xxx tissue development, ...;   
19: axon development;   
**20**: nerve development, neurotransmitter transport, ....

## Reactome Pathway Over-Representation Analysis

Gene sets: The [Reactome pathway database](http://www.reactome.org/).


```{r reactome}
gsfa.Web_ORA_react <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.pathway_Reactome.cutoff_0.95.rds"))

signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_react, enrich_type = "Reactome pathways")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

## KEGG Pathway Over-Representation Analysis

Gene sets: The [KEGG pathway database](http://www.kegg.jp/).

```{r kegg}
gsfa.Web_ORA_kegg <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.pathway_KEGG.cutoff_0.95.rds"))

signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_kegg, enrich_type = "KEGG pathways")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

## Wikipathway Over-Representation Analysis

Gene sets: The [Wikipathway database](http://www.wikipathway.org/).

```{r wiki}
gsfa.Web_ORA_wiki <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.pathway_Wikipathway.cutoff_0.95.rds"))

signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_wiki, enrich_type = "Wikipathways")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

