---
title: "Guided Factor Analysis on LUHMES CROP-seq Data"
subtitle: "-- Deviance residual transformed + 4 covariates corrected"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)

# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
print_signif_tb <- function(signif_num, caption_text){
  if (length(signif_num) < 15){
    signif_num <- c(signif_num, rep(NA, 15 - length(signif_num)))
  }
  signif_num <- t(data.frame(KO = names(signif_num),
                             Num_genes = signif_num,
                             row.names = NULL))
  knitr::kable(rbind(signif_num[, 1:5], signif_num[, 6:10], signif_num[, 11:15]),
                     caption = caption_text) %>%
    kable_styling() %>% scroll_box(width = '100%')
}

wkdir <- "/project2/xinhe/yifan/Factor_analysis/LUHMES/"
gsfa_folder <- "gsfa_output_detect_01/dev_res_normal_mixture/"
guide <- "All"
```

# Single cell expression data

Source:   
[High-throughput single-cell functional elucidation of neurodevelopmental disease-associated genes reveals convergent mechanisms altering neuronal differentiation](https://genome.cshlp.org/content/30/9/1317.full), GEO accession: GSE142078.

Perturbations:    
CRISPR knock-down of 14 autism spectrum disorder (ASD)â€“associated genes (3 gRNAs per gene) + 5 non-targeting gRNAs.

Cells:    
Lund human mesencephalic (LUHMES) neural progenitor cell line.   
Cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout. Quality control resulted in 8708 cells.

Genes:    
**Top 6000 genes ranked by the multinomial deviance statistics were kept.**

Normalization:    
Deviance residual transformation.   
**Batch effect, unique UMI count, library size, and mitochondria percentage were all corrected for. The corrected and scaled expression data were used as input for subsequent factor analysis.**

```{r load meta data, fig.width=9}
metadata <- readRDS(paste0(wkdir, "processed_data/merged_metadata.rds"))
all_KOs.G_mat <- metadata[, 4:18]
KO_names <- colnames(all_KOs.G_mat)
```

Here, our "guide", $G$ matrix, consists of 15 types (14 genes + NTC) of gene-level knock-down conditions across cells.

```{r}
if (guide == "no_NTC"){
  KO_names <- KO_names[KO_names != "Nontargeting"]
  all_KOs.G_mat <- all_KOs.G_mat[, KO_names]
}
nmarkers <- ncol(all_KOs.G_mat)
```

In each case, Gibbs sampling was conducted for 3000 iterations, and the posterior mean estimates were averaged over the last 1000 iterations. SVD Initialization.

```{r svd_tb}
All_KOs.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.restart.PM.rds"))
stopifnot(nrow(All_KOs.gibbs_PM$beta_pm) == nmarkers + 1)
All_KOs.gibbs_res_tb <- make_gibbs_res_tb(All_KOs.gibbs_PM, all_KOs.G_mat, compute_pve = F)
```

# Factor ~ Perturbation Association

## Estimate of Factor ~ Perturbation Associations ($\beta$)

```{r fig.width=9, fig.height=4}
reorder_KO <- c("ADNP", "ARID1B", "ASH1L", "CHD2", "PTEN", "SETD5", "DYRK1A", 
                # "CHD8", "CTNND2", "HDAC5", "MECP2", "MYT1L", "POGZ", "RELN",
                "Nontargeting")
dotplot_beta_PIP(t(All_KOs.gibbs_PM$Gamma_pm), t(All_KOs.gibbs_PM$beta_pm),
                 marker_names = colnames(all_KOs.G_mat), reorder_markers = reorder_KO,
                 inverse_factors = F) +
  coord_flip()
```

## Factor ~ Perturbation Association P Values
```{r fig.width=10, include=FALSE}
heatmap_matrix <- All_KOs.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix[, reorder_KO],
                  factor_annot = All_KOs.gibbs_res_tb$pi)
guided.heatmap_matrix <- heatmap_matrix
gsfa_KO_names <- apply(guided.heatmap_matrix, 1,
                       function(x){ paste(colnames(all_KOs.G_mat)[x < 1e-4], collapse = ",") })
gsfa_KO_names <- paste(1:nrow(guided.heatmap_matrix), gsfa_KO_names, sep = " - ")
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA All KOs (K=20, SVD)")
```

### Beta PIP vs P-Value
```{r fig.width=12, fig.height=5}
beta_pip_matrix <- t(All_KOs.gibbs_PM$Gamma_pm[-nrow(All_KOs.gibbs_PM$Gamma_pm), ])
p1 <- qplot(unlist(beta_pip_matrix), -log10(unlist(heatmap_matrix)),
            xlab = "Beta PIP", ylab = "-log10(Association p-value)",
            main = "20 factors x 15 conditions")
p2 <- qplot(unlist(beta_pip_matrix), -log10(unlist(heatmap_matrix)),
            ylim = c(0, 10),
            xlab = "Beta PIP", ylab = "-log10(Association p-value)",
            main = "Y-axis truncated at 10")
grid.arrange(p1, p2, nrow = 1)
```

### Correlation btw Factors

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = All_KOs.gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Pairwise Correlation in Sample Loadings - Z (GSFA)",
                           label_size = 10)
```

```{r  fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = (All_KOs.gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Binarized Gene Loadings (F.pm > 0.95)",
                           label_size = 10)
```

# Factor Interpretation

## Factor by gene loading

```{r}
expression_file <- "processed_data/deviance_residual.merged_top_6k.corrected_4.scaled.rds"
scaled.gene_exp <- readRDS(paste0(wkdir, expression_file))

feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219576_Run2_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
names(feature.names) <- c("ID", "Name")
genes_indx <- match(rownames(scaled.gene_exp), feature.names$ID)
genes_df <- feature.names[genes_indx, ]
```

```{r}
gene_loadings <- All_KOs.gibbs_PM$W_pm
rownames(gene_loadings) <- genes_df$Name
colnames(gene_loadings) <- paste0("Factor ", 1:20)

gene_pips <- All_KOs.gibbs_PM$F_pm
rownames(gene_pips) <- genes_df$Name
colnames(gene_pips) <- paste0("Factor ", 1:20)
```

Neuron progenitor (NPC): NES, TP53, CDK4, NEUROD1  
Post-mitotic neuron: MAP2, DCX, STMN2  
Glutamatergic neuron (excitatory): SLC17A6  
Cortical neuron: BRN2 (POU3F2)  
Developing striatal neurons: EBF1  
Subtypes of inhibitory neurons: SST  
Subtypes of extitatory neurons: NEFM

```{r list_of_marker_genes}
all_moi_df1 <- data.frame(
  protein_name = c("VIM", "NES", "SOX2", "TP53", "CDK4", "NEUROD1", "MAP2", "DCX", "STMN2", "GAD1", "GAD2", "SLC17A6", "DAT1",
  "FOXG1", "BRN2", "EBF1", "SERTAD4", "TBR1", "NPY", "SST", "DLX2", "DLX5", "SATB2", "CBLN2", "CUX2", "NEFM"),
  gene_name = c("VIM", "NES", "SOX2", "TP53", "CDK4", "NEUROD1", "MAP2", "DCX", "STMN2", "GAD1", "GAD2", "SLC17A6", "SLC6A3",
  "FOXG1", "POU3F2", "EBF1", "SERTAD4", "TBR1", "NPY", "SST", "DLX2", "DLX5", "SATB2", "CBLN2", "CUX2", "NEFM")
)
interest_df1 <- all_moi_df1[all_moi_df1$gene_name %in% genes_df$Name, ]
```

```{r fig.width=9, fig.height=4}
interest_mat <- gene_loadings[interest_df1$gene_name, ]
interest_pip_mat <- gene_pips[interest_df1$gene_name, ]
dotplot_beta_PIP(beta_pip_matrix = t(interest_pip_mat), beta_pm_matrix = t(interest_mat),
                 marker_names = interest_df1$gene_name,
                 exclude_offset = F, color_lgd_title = "Gene loading", inverse_factors = F) +
  coord_flip()
```

SERPINF1 to ITM2C: negative regulators of neuron projection and differentiation    
ZEB2 to DPYSL3: neuron maturation markers  
```{r list_of_neuron_proj_genes}
all_moi_df2 <- data.frame(
  protein_name = c("DPYSL3", "HSPA5", "MAP1B", "CRMP1", "CRABP2", "CCDC88A", "UBE2V2", "MAPT", "NEFL", "MEF2C", "ZEB2",
                   "ITM2C", "CNTN2", "ACAP3", "FKBP4", "GSK3B", "EFNA5", "SERPINF1"),
  gene_name = c("DPYSL3", "HSPA5", "MAP1B", "CRMP1", "CRABP2", "CCDC88A", "UBE2V2", "MAPT", "NEFL", "MEF2C", "ZEB2",
                   "ITM2C", "CNTN2", "ACAP3", "FKBP4", "GSK3B", "EFNA5", "SERPINF1")
)
interest_df2 <- all_moi_df2[all_moi_df2$gene_name %in% genes_df$Name, ]
```

```{r fig.width=9, fig.height=6}
interest_mat <- gene_loadings[interest_df2$gene_name, ]
interest_pip_mat <- gene_pips[interest_df2$gene_name, ]
dotplot_beta_PIP(beta_pip_matrix = t(interest_pip_mat), beta_pm_matrix = t(interest_mat),
                 marker_names = interest_df2$gene_name,
                 exclude_offset = F, color_lgd_title = "Gene loading", inverse_factors = F) +
  coord_flip()
```

## Selected factor GO enrichment

```{r}
enrich_res <- readRDS(paste0(wkdir, gsfa_folder,
                             "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.cutoff_0.95.rds"))
```

```{r fig.height=2.4, fig.width=7}
factor_indx <- 4
terms_of_interest <- c("regulation of ion transmembrane transport",
                       "regulation of trans-synaptic signaling",
                       "axon development",
                       "regulation of neuron projection development")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r fig.height=2.4, fig.width=7}
factor_indx <- 9
terms_of_interest <- c("actin filament organization",
                       "cell fate commitment",
                       "regulation of neuron projection development",
                       "regulation of cell morphogenesis")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r fig.height=2.4, fig.width=7}
factor_indx <- 15
terms_of_interest <- c("neuromuscular process",
                       "regulation of metal ion transport",
                       "regulation of transporter activity",
                       "positive regulation of secretion")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r fig.height=1.8, fig.width=7}
factor_indx <- 16
terms_of_interest <- c("developmental growth involved in morphogenesis",
                       "axon development")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest, 
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

# DEG Interpretation

## Number of DEGs detected under different methods

```{r}
fdr_cutoff <- 0.05
lfsr_cutoff <- 0.05
lfsr_mat <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.restart.lfsr_mat.rds"))
lfsr_signif_num <- colSums(lfsr_mat < lfsr_cutoff)
# print_signif_tb(lfsr_signif_num, paste0("GSFA, # of genes that passed LFSR cutoff of 0.05:"))
```

```{r deseq}
library(DESeq2)
deseq_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/DESeq2/dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  res <- readRDS(fname)
  res <- as.data.frame(res@listData, row.names = res@rownames)
  deseq_list[[m]] <- res
}
deseq_signif_counts <- sapply(deseq_list, function(x){filter(x, padj < fdr_cutoff) %>% nrow()})
```

```{r edger}
edger_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/edgeR_qlf/dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  edger_list[[m]] <- readRDS(fname)
}
edger_signif_counts <- sapply(edger_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r mast}
mast_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/MAST/dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  mast_list[[m]] <- readRDS(fname)
}
mast_list_signif_counts <- sapply(mast_list, function(x){filter(x, fdr < fdr_cutoff) %>% nrow()})
```

```{r dge_comparison_merge}
dge_comparison_df <- data.frame(Perturbation = names(lfsr_signif_num),
                                GSFA = lfsr_signif_num,
                                edgeR = edger_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_list_signif_counts)
                                # t.test = DE_signif_counts)
```

```{r fig.width=12, fig.height=5.5}
dge_plot_df <- reshape2::melt(dge_comparison_df, id.var = "Perturbation",
                              variable.name = "Method", value.name = "Num_DEGs")
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Knock-Out Target",
       y = "Number of DEGs + 1",
       title = "Number of DEGs detected using different methods\n(LFSR or FDR cutoff at 0.05)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

## Perturbation effect on gene markers

```{r}
interest_df <- rbind(interest_df1, interest_df2)
lfsr_mat <- readRDS(paste0(wkdir, gsfa_folder, "All.gibbs_obj_k20.svd.restart.lfsr_mat.rds"))
stopifnot(rownames(lfsr_mat) == genes_df$ID)
rownames(lfsr_mat) <- genes_df$Name

lfsr_cutoff <- 0.05
num_signif_genes <- colSums(lfsr_mat < lfsr_cutoff)
signif_markers <- names(num_signif_genes)[which(num_signif_genes > 0)]

effect_size_mat <- matrix(nrow = nrow(interest_df),
                          ncol = length(signif_markers))
rownames(effect_size_mat) <- interest_df$protein_name
colnames(effect_size_mat) <- signif_markers

beta1_mat <- All_KOs.gibbs_PM$beta_pm
rownames(beta1_mat) <- c(colnames(lfsr_mat), "offset")
gamma1_mat <- All_KOs.gibbs_PM$Gamma_pm
rownames(gamma1_mat) <- c(colnames(lfsr_mat), "offset")

for (m in signif_markers){
  effect_size_mat[, m] <-
    colSums(beta1_mat[m, ] * t(gene_loadings[interest_df$gene_name, ]))
}
```

```{r dotplot_gene_effect1, fig.width=5, fig.height=4}
selected_lfsr_mat <- lfsr_mat[interest_df1$gene_name, signif_markers]
rownames(selected_lfsr_mat) <- interest_df1$protein_name
dotplot_effectsize(effect_size_mat[interest_df1$protein_name, ],
                   selected_lfsr_mat,
                   reorder_markers = signif_markers,
                   color_lgd_title = "Summarized effect") +
  theme(axis.text.y = element_text(hjust = 1))
```

SERPINF1 to ITM2C: negative regulators of neuron projection and differentiation    

ZEB2 to DPYSL3: neuron maturation markers  
```{r dotplot_gene_effect2, fig.width=5, fig.height=6}
selected_lfsr_mat <- lfsr_mat[interest_df2$gene_name, signif_markers]
dotplot_effectsize(effect_size_mat[interest_df2$protein_name, ],
                   selected_lfsr_mat,
                   reorder_markers = signif_markers,
                   color_lgd_title = "Summarized effect") +
  theme(axis.text.y = element_text(hjust = 1))
```

## DEG GO enrichment

```{r}
enrich_folder <- paste0(wkdir, gsfa_folder, "WebGestalt_ORA/")
signif_term_clusters <- readRDS(paste0(enrich_folder, "go_bp.by_marker.signif_term_clusters.rds"))
signif_term_df <- signif_term_clusters$signif_term_df

abs_FC_colormap <- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c("grey95", "#77d183", "#255566"))
```

```{r}
enrich_res <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.by_marker.lfsr_0.05.rds"))
targets <- names(enrich_res)
enrich_table <- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) <- targets
for (i in 1:ncol(enrich_table)){
  m <- colnames(enrich_table)[i]
  enrich_df <- enrich_res[[m]]
  enrich_table[enrich_df$geneSet, i] <- enrich_df$enrichmentRatio
}
signif_term_df$degree <- rowSums(!is.na(enrich_table))

ordered_signif_term_df <- signif_term_df %>%
  filter(degree > 0, size >= 20) %>%
  arrange(group, -degree)
ordered_enrich_table <- enrich_table[ordered_signif_term_df$geneSet, ]
rownames(ordered_enrich_table) <- ordered_signif_term_df$description
```

```{r GO_term_heatmap, fig.height=12, fig.width=9}
interest_enrich_table <- ordered_enrich_table
interest_enrich_table[is.na(interest_enrich_table)] <- 0
abs_FC_colormap <- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c("grey95", "#77d183", "#255566"))
map <- Heatmap(interest_enrich_table,
               name = "Fold of Enrichment",
               col = abs_FC_colormap,
               row_title = NULL, column_title = NULL,
               cluster_rows = T, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               width = unit(10, "cm"))
draw(map, heatmap_legend_side = "right")
```

# Gene Set Enrichment Analysis by Factor

Target genes: Genes w/ non-zero loadings in each factor (PIP cutoff at 0.95);    
Backgroud genes: all 6000 genes used in factor analysis;    
Statistical test: hypergeometric test (over-representation test);    
Only GO terms/pathways that satisfy fold change $\geq$ 2 and test FDR $<$ 0.05 are shown below.

```{r}
feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219576_Run2_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
names(feature.names) <- c("ID", "Name")
```

## GO Slim Over-Representation Analysis

Gene sets: [Gene ontology](http://www.geneontology.org/) "Biological Process" (non-redundant).

```{r go_bp}
gsfa.Web_ORA_go_bp <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.cutoff_0.95.rds"))
signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_go_bp, enrich_type = "GO terms",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Factors of interest

4, 9, 15, 16

## Reactome Pathway Over-Representation Analysis

Gene sets: The [Reactome pathway database](http://www.reactome.org/).

```{r reactome}
gsfa.Web_ORA_react <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.Reactome.cutoff_0.95.rds"))

signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_react, enrich_type = "Reactome pathways",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```


# Gene Set Enrichment Analysis by Perturbation

Target genes: Genes w/ LFSR < 0.05 under each perturbation marker gene;    
Backgroud genes: all 6000 genes used in factor analysis;    
Statistical test: hypergeometric test (over-representation test);    
Only GO terms/pathways that satisfy fold change $\geq$ 2 and test FDR $<$ 0.05 are shown below.

## GO Slim Over-Representation Analysis

Gene sets: [Gene ontology](http://www.geneontology.org/) "Biological Process" (non-redundant).

```{r go_bp_by_marker}
Web_ORA_go_bp.by_marker <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.by_marker.lfsr_0.05.rds"))
signif_num <- print_enrich_ORA_tb(Web_ORA_go_bp.by_marker,
                                  enrich_type = "GO terms", list_type = "per_marker",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Marker = names(Web_ORA_go_bp.by_marker),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(signif_num.df, caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

## Reactome Pathway Over-Representation Analysis

Gene sets: The [Reactome pathway database](http://www.reactome.org/).

```{r reactome_by_marker}
Web_ORA_react.by_marker <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.Reactome.by_marker.lfsr_0.05.rds"))

signif_num <- print_enrich_ORA_tb(Web_ORA_react.by_marker,
                                  enrich_type = "Reactome pathways", list_type = "per_marker",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Marker = names(Web_ORA_react.by_marker),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(signif_num.df, caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```
