---
title: "Pseudotime Trajectory Analysis of LUHMES CROP-seq Data"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: no
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(Seurat)
library(monocle)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 13),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)

# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
wkdir <- "/project2/xinhe/yifan/Factor_analysis/LUHMES/"
```

## Single cell expression data

Source:   
[High-throughput single-cell functional elucidation of neurodevelopmental disease-associated genes reveals convergent mechanisms altering neuronal differentiation](https://genome.cshlp.org/content/30/9/1317.full), GEO accession: GSE142078.

Perturbations:    
CRISPR knock-down of 14 autism spectrum disorder (ASD)â€“associated genes (3 gRNAs per gene) + 5 non-targeting gRNAs.

Cells:    
Lund human mesencephalic (LUHMES) neural progenitor cell line.   
Cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout.

```{r}
combined_obj <- readRDS(paste0(wkdir, "processed_data/seurat_obj.merged_scaled_detect_01.corrected_new.rds"))
expression_matrix <- combined_obj@assays$RNA@scale.data

feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219577_Run3_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
feature.names <- feature.names[match(rownames(expression_matrix), feature.names$V1), ]
names(feature.names) <- c("ID", "gene_short_name")
rownames(feature.names) <- feature.names$ID

stopifnot(identical(rownames(feature.names), rownames(expression_matrix)))
```

## Monocle Pseudotime Trajectory

We attempted to reproduce the pseudotime trajectory of cells as reported in the [original paper](https://pubmed.ncbi.nlm.nih.gov/32887689/#&gid=article-figures&pid=figure-3-uid-2).
```{r}
cds <- readRDS(paste0(wkdir, "processed_data/monocle_pseudotime.merged_scaled_detect_01.corrected_new.rds"))
traj_plot <- plot_cell_trajectory(cds, color_by = "State", cell_size = 0.6)
```

```{r fig.height=5}
state_cols <- c(RColorBrewer::brewer.pal(6, "YlOrRd")[2:6], RColorBrewer::brewer.pal(9, "RdPu")[7:9], "#000000")
state_cols <- state_cols[length(state_cols):1]
ggplot(traj_plot$data,
       aes_string(x = "data_dim_1", y = "data_dim_2", color = "sample_state")) +
  geom_point(size = 0.3) +
  scale_color_manual(values = state_cols) +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  labs(x = "Component 1", y = "Component 2", 
       color = "Pseudotime state",
       title = "Pseudotime Trajectory (More Differentiated <-- Less Differentiated)")
```

### Marker Gene Expression in Pseudotime

Neuronal markers (MAP2 and DCX) increase along the pseudotime trajectory, whereas progenitor markers (TP53 and CDK4) decrease, consistent with Fig 3B of original paper.

```{r fig.width=12, fig.height=8}
marker_genes <- c("MAP2", "DCX", "TP53", "CDK4")
plot.lst <- list()
for (m in marker_genes){
  traj_plot$data[[m]] <-
    expression_matrix[feature.names$ID[match(m, feature.names$gene_short_name)], ]
  p <- ggplot(traj_plot$data,
              aes_string(x = "data_dim_1", y = "data_dim_2", color = m)) +
    geom_point(size = 0.3) +
    scale_color_gradient2(low = scales::muted("blue"), high = scales::muted("red"), mid = "grey90") +
    # scale_color_gradientn(colors = c("grey90", "purple3")) +
    labs(x = "Component 1", y = "Component 2", title = m,
         color = "Scaled expression")
  plot.lst[[m]] <- p
}
args <- c(plot.lst, list(nrow = 2))
do.call(grid.arrange, args)
```

## Association btw Pseudotime States and GSFA Factors

```{r}
All_KOs.gibbs_PM <- readRDS(paste0(wkdir, "gsfa_output_detect_01/dev_res_normal_mixture/",
                                   "All.gibbs_obj_k20.svd.restart.PM.rds"))
time_meta <- data.frame(state = traj_plot$data$State)
corr_tb <- make_gibbs_res_tb(All_KOs.gibbs_PM, time_meta, compute_pve = F)
corr_tb <- corr_tb %>%
  rename(Factor = index, Density = pi, Coefficient = `beta_reg-state`, P_value = `pval-state`) %>%
  arrange(P_value)
corr_tb$Density <- signif(corr_tb$Density, digits = 2)
corr_tb$Coefficient <- signif(corr_tb$Coefficient, digits = 3)
corr_tb$P_value <- format(corr_tb$P_value, digits = 3)
kable(corr_tb) %>% kable_styling() %>% scroll_box(height = "500px", width = "100%")
```

## Over/Under-Representation of gRNA Targets in Different Pseudotime States

```{r state_chisq_test}
combined_meta <- traj_plot$data
targets <- colnames(combined_meta)[8:22]
chisq_stdres <- matrix(nrow = length(levels(combined_meta$State)),
                       ncol = length(targets))
colnames(chisq_stdres) <- targets
rownames(chisq_stdres) <- paste0("State ", 1:nrow(chisq_stdres))
chisq_pval <- chisq_stdres
for (m in targets){
  test <- chisq.test(table(factor(combined_meta[[m]]), combined_meta$State))
  chisq_stdres[, m] <- test$stdres[2, ]
  for (k in 1:nrow(chisq_stdres)){
    chisq_pval[k, m] <- chisq.test(table(factor(combined_meta[[m]]),
                                         combined_meta$State == k))$p.value
  }
}
```

```{r fig.height=4.5, fig.width=7}
Heatmap(t(-log10(chisq_pval) * sign(chisq_stdres)),
        name = "chi-square polarized\n-log10(p value)",
        cluster_rows = F, cluster_columns = F)
```

Pairs of associations that passed chi-square test FDR < 0.05:

```{r}
chisq_pval_adj <- p.adjust(unlist(chisq_pval), method = "fdr")
cat(sum(chisq_pval_adj < 0.05))
cat("\n\n")
```

```{r}
chisq_pval_adj <- matrix(chisq_pval_adj, nrow = nrow(chisq_pval))
rownames(chisq_pval_adj) <- rownames(chisq_pval)
colnames(chisq_pval_adj) <- colnames(chisq_pval)
signif_guides <- colnames(chisq_pval_adj)[colSums(chisq_pval_adj < 0.05) > 0]
cat("Perturbations with significant enrichment or depletion in states:\n")
cat(paste(signif_guides, collapse = ", "))
```

