---
title: "Guided Factor Analysis on LUHMES CROP-seq Data"
subtitle: "-- Deviance residual transformed + normal-mixture prior + by-stage estimation"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(Seurat)
library(monocle)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)

# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
print_signif_tb <- function(signif_num, caption_text){
  if (length(signif_num) < 15){
    signif_num <- c(signif_num, rep(NA, 15 - length(signif_num)))
  }
  signif_num <- t(data.frame(KO = names(signif_num),
                             Num_genes = signif_num,
                             row.names = NULL))
  knitr::kable(rbind(signif_num[, 1:5], signif_num[, 6:10], signif_num[, 11:15]),
                     caption = caption_text) %>%
    kable_styling() %>% scroll_box(width = '100%')
}

wkdir <- "/project2/xinhe/yifan/Factor_analysis/LUHMES/"
gsfa_folder <- "gsfa_output_detect_01/by_groups_normal_mixture/"
guide <- "All"
```

# Single cell expression data

Source:   
[High-throughput single-cell functional elucidation of neurodevelopmental disease-associated genes reveals convergent mechanisms altering neuronal differentiation](https://genome.cshlp.org/content/30/9/1317.full), GEO accession: GSE142078.

Perturbations:    
CRISPR knock-down of 14 autism spectrum disorder (ASD)â€“associated genes (3 gRNAs per gene) + 5 non-targeting gRNAs.

Cells:    
Lund human mesencephalic (LUHMES) neural progenitor cell line.   
Cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout. Quality control resulted in 8708 cells.

Genes:    
**Top 6000 genes ranked by the multinomial deviance statistics were kept.**

Normalization:    
Deviance residual transformation.   
**Batch effect, unique UMI count, library size, and mitochondria percentage were all corrected for. The corrected and scaled expression data were used as input for subsequent factor analysis.**

```{r load meta data, fig.width=9}
metadata <- readRDS(paste0(wkdir, "processed_data/merged_metadata.rds"))
all_KOs.G_mat <- metadata[, 4:18]
KO_names <- colnames(all_KOs.G_mat)
```

Here, our "guide", $G$ matrix, consists of 15 types (14 genes + NTC) of gene-level knock-down conditions across cells.

```{r}
if (guide == "no_NTC"){
  KO_names <- KO_names[KO_names != "Nontargeting"]
  all_KOs.G_mat <- all_KOs.G_mat[, KO_names]
}
nmarkers <- ncol(all_KOs.G_mat)
```

In each case, Gibbs sampling was conducted for 3000 iterations, and the posterior mean estimates were averaged over the last 1000 iterations. SVD Initialization.

```{r svd_tb}
All_KOs.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.PM.rds"))
stopifnot(nrow(All_KOs.gibbs_PM$beta_pm) == nmarkers + 1)
```

## Monocle Pseudotime Trajectory

We attempted to reproduce the pseudotime trajectory of cells as reported in the [original paper](https://pubmed.ncbi.nlm.nih.gov/32887689/#&gid=article-figures&pid=figure-3-uid-2).
```{r}
cds <- readRDS(paste0(wkdir, "processed_data/monocle_pseudotime.merged_scaled_detect_01.corrected_new.rds"))
traj_plot <- plot_cell_trajectory(cds, color_by = "State", cell_size = 0.6) 
```

```{r fig.height=5}
state_cols <- c(RColorBrewer::brewer.pal(6, "YlOrRd")[2:6], RColorBrewer::brewer.pal(9, "RdPu")[7:9], "#000000")
state_cols <- state_cols[length(state_cols):1]
ggplot(traj_plot$data,
       aes_string(x = "data_dim_1", y = "data_dim_2", color = "sample_state")) +
  geom_point(size = 0.3) +
  scale_color_manual(values = state_cols) +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  labs(x = "Component 1", y = "Component 2", 
       color = "Pseudotime state",
       title = "Pseudotime Trajectory (More Differentiated <-- Less Differentiated)")
```

## Marker Gene Expression in Pseudotime

Neuronal markers (MAP2 and DCX) increase along the pseudotime trajectory, whereas progenitor markers (TP53 and CDK4) decrease, consistent with Fig 3B of original paper.

Neuron progenitor (NPC): TP53, CDK4, NEUROD1, NES?  
Post-mitotic neuron: MAP2, DCX, STMN2  

```{r}
combined_obj <- readRDS(paste0(wkdir, "processed_data/seurat_obj.merged_scaled_detect_01.corrected_new.rds"))
expression_matrix <- combined_obj@assays$RNA@scale.data

feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219576_Run2_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
feature.names <- feature.names[match(rownames(expression_matrix), feature.names$V1), ]
names(feature.names) <- c("ID", "Name")
rownames(feature.names) <- feature.names$ID

stopifnot(identical(rownames(feature.names), rownames(expression_matrix)))
```

```{r fig.width=12, fig.height=16}
marker_genes <- c("TP53", "CDK4", "NEUROD1", "NES", "MAP2", "DCX", "STMN2")
plot.lst <- list()
for (m in marker_genes){
  traj_plot$data[[m]] <-
    expression_matrix[feature.names$ID[match(m, feature.names$Name)], ]
  p <- ggplot(traj_plot$data,
              aes_string(x = "data_dim_1", y = "data_dim_2", color = m)) +
    geom_point(size = 0.3) +
    scale_color_gradient2(low = scales::muted("blue"), high = scales::muted("red"), mid = "grey90") +
    # scale_color_gradientn(colors = c("grey90", "purple3")) +
    labs(x = "Component 1", y = "Component 2", title = m,
         color = "Scaled expression")
  plot.lst[[m]] <- p
}
args <- c(plot.lst, list(ncol = 2))
do.call(grid.arrange, args)
```

Glutamatergic neuron (excitatory): SLC17A6  
Cortical neuron: BRN2 (POU3F2)  
Developing striatal neurons: EBF1  
Subtypes of inhibitory neurons: SST  
Subtypes of extitatory neurons: NEFM

```{r fig.width=12, fig.height=8}
marker_genes <- c("SLC17A6", "POU3F2", "EBF1", "NEFM")
plot.lst <- list()
for (m in marker_genes){
  traj_plot$data[[m]] <-
    expression_matrix[feature.names$ID[match(m, feature.names$Name)], ]
  p <- ggplot(traj_plot$data,
              aes_string(x = "data_dim_1", y = "data_dim_2", color = m)) +
    geom_point(size = 0.3) +
    scale_color_gradient2(low = scales::muted("blue"), high = scales::muted("red"), mid = "grey90") +
    # scale_color_gradientn(colors = c("grey90", "purple3")) +
    labs(x = "Component 1", y = "Component 2", title = m,
         color = "Scaled expression")
  plot.lst[[m]] <- p
}
args <- c(plot.lst, list(ncol = 2))
do.call(grid.arrange, args)
```

**We separate the cells into 2 groups:**

Early stage: 2523 cells in pseudotime state 1 - 3;  
Late stage: 6185 cells in pseudotime satte 4 - 9.

We modified GSFA so that the associations ($\beta$) between factors and perturbations are estimated for early and late stage cells separately.

# Factor ~ Perturbation Association

## Early stage cells

```{r fig.width=8.5, fig.height=4}
reorder_KO <- c("ADNP", "CHD2", "PTEN", "SETD5", "ASH1L", "ARID1B", "DYRK1A", "Nontargeting")[8:1]
                # "CHD8", "CTNND2", "HDAC5", "MECP2", "MYT1L", "POGZ", "RELN"
dotplot_beta_PIP(t(All_KOs.gibbs_PM$Gamma0_pm), t(All_KOs.gibbs_PM$beta0_pm),
                 marker_names = colnames(all_KOs.G_mat), reorder_markers = reorder_KO,
                 inverse_factors = F) +
  coord_flip() +
  labs(title = "Early Stage") +
  theme(plot.title = element_text(size = 14, hjust = 0.5))
```

## Late stage cells

```{r fig.width=8.5, fig.height=4}
reorder_KO <- c("ADNP", "CHD2", "PTEN", "SETD5", "ASH1L", "ARID1B", "DYRK1A", "Nontargeting")[8:1]
                # "CHD8", "CTNND2", "HDAC5", "MECP2", "MYT1L", "POGZ", "RELN"
dotplot_beta_PIP(t(All_KOs.gibbs_PM$Gamma1_pm), t(All_KOs.gibbs_PM$beta1_pm),
                 marker_names = colnames(all_KOs.G_mat), reorder_markers = reorder_KO,
                 inverse_factors = F) +
  coord_flip() +
  labs(title = "Late Stage") +
  theme(plot.title = element_text(size = 14, hjust = 0.5))
```

## Correlation within Factors

```{r fig.width=7, fig.height=5}
plot_pairwise.corr_heatmap(input_mat_1 = All_KOs.gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Sample Loadings - Z (GSFA)",
                           label_size = 10)
```

```{r  fig.width=7, fig.height=5, eval=F, include=F}
plot_pairwise.corr_heatmap(input_mat_1 = (All_KOs.gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Binarized Gene Loadings (F.pm > 0.95)",
                           label_size = 10)
```

### Correlation with factors inferred using all cells

```{r fig.width=7, fig.height=5}
all_cells.gibbs_PM <- readRDS(paste0(wkdir, "gsfa_output_detect_01/dev_res_normal_mixture/",
                                     guide, ".gibbs_obj_k20.svd.restart.PM.rds"))
plot_pairwise.corr_heatmap(input_mat_1 = All_KOs.gibbs_PM$Z_pm,
                           input_mat_2 = all_cells.gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Sample Loadings - Z (all cells inference)",
                           name_2 = "Sample Loadings - Z (by stage inference)",
                           label_size = 10)
```

# Factor Interpretation

## Factor by gene loading

```{r}
expression_file <- "processed_data/deviance_residual.merged_top_6k.corrected_4.scaled.rds"
scaled.gene_exp <- readRDS(paste0(wkdir, expression_file))

feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219576_Run2_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
names(feature.names) <- c("ID", "Name")
genes_indx <- match(rownames(scaled.gene_exp), feature.names$ID)
genes_df <- feature.names[genes_indx, ]
```

```{r}
gene_loadings <- All_KOs.gibbs_PM$W_pm
rownames(gene_loadings) <- genes_df$Name
colnames(gene_loadings) <- paste0("Factor ", 1:ncol(gene_loadings))

gene_pips <- All_KOs.gibbs_PM$F_pm
rownames(gene_pips) <- genes_df$Name
colnames(gene_pips) <- paste0("Factor ", 1:ncol(gene_pips))
```

Cell differentiation markers and neuronal cell type markers:

```{r}
interest_df <- readRDS(paste0(wkdir, "processed_data/selected_neuronal_markers.rds"))
interest_df <- interest_df[interest_df$gene_name %in% genes_df$Name, ]
rownames(interest_df) <- NULL
knitr::kable(interest_df) %>%
    kable_styling() %>% scroll_box(width = '100%', height = '400px')
```

```{r fig.width=8, fig.height=6.5}
interest_mat <- gene_loadings[interest_df$gene_name, ]
interest_pip_mat <- gene_pips[interest_df$gene_name, ]
dotplot_beta_PIP(beta_pip_matrix = t(interest_pip_mat),
                 beta_pm_matrix = t(interest_mat),
                 marker_names = interest_df$gene_name,
                 reorder_markers = interest_df$gene_name[nrow(interest_df):1],
                 exclude_offset = F, color_lgd_title = "Gene loading", inverse_factors = F) +
  coord_flip()
```

## Selected factor GO enrichment

```{r}
enrich_res <- readRDS(paste0(wkdir, gsfa_folder,
                             "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.cutoff_0.95.rds"))
```

```{r fig.height=2.4, fig.width=7}
factor_indx <- 4
terms_of_interest <- c("regulation of metal ion transport",
                       "regulation of trans-synaptic signaling",
                       "axon development",
                       "regulation of neuron projection development")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r fig.height=2.4, fig.width=7}
factor_indx <- 9
terms_of_interest <- c("actin filament-based movement",
                       "cell fate commitment",
                       "sodium ion transport",
                       "axon development")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

```{r fig.height=2, fig.width=7}
factor_indx <- 15
terms_of_interest <- c("monoamine transport",
                       "muscle organ development")
barplot_top_enrich_terms(enrich_res[[factor_indx]], terms_of_interest = terms_of_interest,
                         size_cutoff = 10, str_wrap_length = 35, pval_max = 8, FC_max = 6) +
  labs(title = paste0("Factor ", factor_indx))
```

# DEG Interpretation

## Number of DEGs detected under different methods

### Early stage cells

Number of genes that passed GSFA LFSR < 0.05 under each perturbation:
```{r}
lfsr_cutoff <- 0.05
fdr_cutoff <- 0.05
lfsr_mat0 <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.lfsr_mat.group0.rds"))
lfsr_signif_num <- colSums(lfsr_mat0 < lfsr_cutoff)
# print_signif_tb(lfsr_signif_num, paste0("GSFA, # of genes that passed LFSR cutoff of 0.05:"))
```

```{r deseq_0}
library(DESeq2)
deseq_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/DESeq2/by_early_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  res <- readRDS(fname)
  res <- as.data.frame(res@listData, row.names = res@rownames)
  deseq_list[[m]] <- res
}
deseq_signif_counts <- sapply(deseq_list, function(x){filter(x, padj < fdr_cutoff) %>% nrow()})
```

```{r edger_0}
edger_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/edgeR_qlf/by_early_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  edger_list[[m]] <- readRDS(fname)
}
edger_signif_counts <- sapply(edger_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r mast_0}
mast_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/MAST/by_early_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  mast_list[[m]] <- readRDS(fname)
}
mast_list_signif_counts <- sapply(mast_list, function(x){filter(x, fdr < fdr_cutoff) %>% nrow()})
```

```{r}
scmageck_res <- readRDS(paste0(wkdir, "scmageck/scmageck_lr.LUHMES_early.dev_res_top_6k.rds"))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == "NegCtrl"] <- "Nontargeting"
scmageck_signif_counts <- colSums(scmageck_res$fdr[, KO_names] < fdr_cutoff)
```

```{r dge_comparison_merge_0}
dge_comparison_df <- data.frame(Perturbation = KO_names,
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                edgeR = edger_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_list_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == "Nontargeting"] <- "NegCtrl"
```

```{r fig.width=12, fig.height=5.5}
dge_plot_df <- reshape2::melt(dge_comparison_df, id.var = "Perturbation",
                              variable.name = "Method", value.name = "Num_DEGs")
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation,
                                   levels = c("NegCtrl", KO_names[KO_names!="Nontargeting"]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Knock-Out Target",
       y = "Number of DEGs + 1",
       title = "Number of DEGs detected using different methods\n(Early stage cells, LFSR or FDR cutoff at 0.05)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

### Late stage cells

Number of genes that passed GSFA LFSR < 0.05 under each perturbation:
```{r}
lfsr_cutoff <- 0.05
lfsr_mat1 <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.lfsr_mat.group1.rds"))
lfsr_signif_num <- colSums(lfsr_mat1 < lfsr_cutoff)
# print_signif_tb(lfsr_signif_num, paste0("GSFA, # of genes that passed LFSR cutoff of 0.05:"))
```

```{r deseq_1}
library(DESeq2)
deseq_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/DESeq2/by_late_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  res <- readRDS(fname)
  res <- as.data.frame(res@listData, row.names = res@rownames)
  deseq_list[[m]] <- res
}
deseq_signif_counts <- sapply(deseq_list, function(x){filter(x, padj < fdr_cutoff) %>% nrow()})
```

```{r edger_1}
edger_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/edgeR_qlf/by_late_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  edger_list[[m]] <- readRDS(fname)
}
edger_signif_counts <- sapply(edger_list, function(x){filter(x, FDR < fdr_cutoff) %>% nrow()})
```

```{r mast_1}
mast_list <- list()
for (m in KO_names){
  fname <- paste0(wkdir, "processed_data/MAST/by_late_dev_top6k/gRNA_", m, ".dev_res_top6k.rds")
  mast_list[[m]] <- readRDS(fname)
}
mast_list_signif_counts <- sapply(mast_list, function(x){filter(x, fdr < fdr_cutoff) %>% nrow()})
```

```{r scmageck_1}
scmageck_res <- readRDS(paste0(wkdir, "scmageck/scmageck_lr.LUHMES_late.dev_res_top_6k.rds"))
colnames(scmageck_res$fdr)[colnames(scmageck_res$fdr) == "NegCtrl"] <- "Nontargeting"
scmageck_signif_counts <- colSums(scmageck_res$fdr[, KO_names] < fdr_cutoff)
```

```{r dge_comparison_merge_1}
dge_comparison_df <- data.frame(Perturbation = KO_names,
                                GSFA = lfsr_signif_num,
                                scMAGeCK = scmageck_signif_counts,
                                edgeR = edger_signif_counts,
                                DESeq2 = deseq_signif_counts,
                                MAST = mast_list_signif_counts)
dge_comparison_df$Perturbation[dge_comparison_df$Perturbation == "Nontargeting"] <- "NegCtrl"
```

```{r fig.width=12, fig.height=5.5}
dge_plot_df <- reshape2::melt(dge_comparison_df, id.var = "Perturbation",
                              variable.name = "Method", value.name = "Num_DEGs")
dge_plot_df$Perturbation <- factor(dge_plot_df$Perturbation,
                                   levels = c("NegCtrl", KO_names[KO_names!="Nontargeting"]))
ggplot(dge_plot_df, aes(x = Perturbation, y = Num_DEGs+1, fill = Method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = Num_DEGs), position=position_dodge(width=0.9), vjust=-0.25) +
  scale_y_log10() +
  scale_fill_brewer(palette = "Set2") +
  labs(x = "Knock-Out Target",
       y = "Number of DEGs + 1",
       title = "Number of DEGs detected using different methods\n(Late stage cells, LFSR or FDR cutoff at 0.05)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        legend.position = "bottom",
        legend.text = element_text(size = 13))
```

## Perturbation effect on gene markers

```{r}
complexplot_gene_perturbation <- function(genes_df, interest_df,
                                          signif_markers = NULL, lfsr_cutoff = 0.05,
                                          lfsr_mat, beta_mat, W_pm){
  interest_df <- interest_df[interest_df$gene_name %in% genes_df$Name, ]
  interest_df$type <- factor(interest_df$type, levels = unique(interest_df$type))
  stopifnot(rownames(lfsr_mat) == genes_df$ID)
  rownames(lfsr_mat) <- genes_df$Name
  if (is.null(signif_markers)){
    num_signif_genes <- colSums(lfsr_mat < lfsr_cutoff)
    signif_markers <- names(num_signif_genes)[which(num_signif_genes > 0)]
  }
  rownames(beta_mat) <- c(colnames(lfsr_mat), "offset")
  rownames(W_pm) <- genes_df$Name
  
  effect_size_mat <- matrix(nrow = nrow(interest_df),
                            ncol = length(signif_markers))
  rownames(effect_size_mat) <- interest_df$gene_name
  colnames(effect_size_mat) <- signif_markers
  for (m in signif_markers){
    effect_size_mat[, m] <-
      colSums(beta_mat[m, ] * t(W_pm[interest_df$gene_name, ]))
  }
  selected_lfsr_mat <- lfsr_mat[interest_df$gene_name, signif_markers]
  binned_size_mat <- matrix(0.6, 
                            nrow = nrow(selected_lfsr_mat),
                            ncol = ncol(selected_lfsr_mat))
  rownames(binned_size_mat) <- rownames(selected_lfsr_mat)
  colnames(binned_size_mat) <- colnames(selected_lfsr_mat)
  binned_size_mat[selected_lfsr_mat <= 0.05] <- 1
  binned_size_mat[selected_lfsr_mat > 0.25] <- 0.2
  
  lgd_list <- list()
  col_fun <- circlize::colorRamp2(breaks = seq(-0.2, 0.2, 0.1),
                                  colors = c("blue3", "blue", "grey90", "red", "red3"))
  lgd_list[["effectsize"]] <- Legend(title = "Summarized effect",
                                     at = seq(-0.2, 0.2, 0.1),
                                     col_fun = col_fun)
  
  lfsr_tic_values <- c(0.2, 0.6, 1)
  lfsr_tic_labels <- c("> 0.25", "0.05 - 0.25", "0 - 0.05")
  
  lgd_list[["LFSR"]] <- 
    Legend(title = "LFSR",
           labels = lfsr_tic_labels,
           # labels_gp = gpar(fontsize = 14),
           grid_height = unit(6, "mm"),
           grid_width = unit(6, "mm"),
           graphics = list(
             function(x, y, w, h) grid.circle(x, y, r = (lfsr_tic_values[1] + 0.2) * unit(2, "mm"),
                                              gp = gpar(fill = "black")),
             function(x, y, w, h) grid.circle(x, y, r = (lfsr_tic_values[2] + 0.2) * unit(2, "mm"),
                                              gp = gpar(fill = "black")),
             function(x, y, w, h) grid.circle(x, y, r = (lfsr_tic_values[3] + 0.2) * unit(2, "mm"),
                                              gp = gpar(fill = "black"))
           ))
  
  marker_colormap <- structure(RColorBrewer::brewer.pal(length(levels(interest_df$type)), "Accent"),
                               names = levels(interest_df$type))
  lgd_list[["Marker"]] <-  Legend(title = "Marker annotation",
                                labels = levels(interest_df$type),
                                at = levels(interest_df$type),
                                legend_gp = gpar(fill = marker_colormap))
  right_annot <- rowAnnotation(Marker = interest_df$type,
                               col = list(Marker = marker_colormap),
                               annotation_legend_param = list(
                                 Marker = list(
                                   title = "Marker annotation",
                                   at = levels(interest_df$type),
                                   labels = levels(interest_df$type)
                                 )
                               ),
                               show_annotation_name = F,
                               show_legend = F)
  
  map1 <- Heatmap(effect_size_mat,
                  name = "Summarized effect",
                  col = col_fun,
                  rect_gp = gpar(type = "none"),
                  cell_fun = function(j, i, x, y, width, height, fill) {
                    grid.rect(x = x, y = y, width = width, height = height, 
                              gp = gpar(col = NA, fill = NA))
                    grid.circle(x = x, y = y,
                                r = (binned_size_mat[i, j] + 0.2) * unit(2, "mm"),
                                gp = gpar(fill = col_fun(effect_size_mat[i, j]), col = NA))
                  },
                  border_gp = gpar(col = "black"),
                  row_title = "Marker genes",
                  column_title = "Perturbations",
                  cluster_rows = F, cluster_columns = F,
                  right_annotation = right_annot,
                  show_heatmap_legend = F,
                  row_names_gp = gpar(fontsize = 10.5),
                  column_names_rot = 45,
                  column_names_side = "top",
                  column_title_side = "bottom")
  draw(map1, annotation_legend_list = lgd_list)
}
```

Cell differentiation markers and neuronal cell type markers:

```{r}
interest_df <- readRDS(paste0(wkdir, "processed_data/selected_neuronal_markers.rds"))
interest_df <- interest_df[interest_df$gene_name %in% genes_df$Name, ]
rownames(interest_df) <- NULL
knitr::kable(interest_df) %>%
    kable_styling() %>% scroll_box(width = '100%', height = '400px')
```

### Early stage cells

```{r fig.width=6.5, fig.height=6}
complexplot_gene_perturbation(genes_df, interest_df,
                            signif_markers = c("ADNP", "CHD2", "PTEN", "SETD5", "ASH1L", "ARID1B", "DYRK1A"),
                            lfsr_mat = lfsr_mat0,
                            beta_mat = All_KOs.gibbs_PM$beta0_pm,
                            W_pm = All_KOs.gibbs_PM$W_pm)
```

### Late stage cells

```{r fig.width=6.5, fig.height=6}
complexplot_gene_perturbation(genes_df, interest_df,
                            signif_markers = c("ADNP", "CHD2", "PTEN", "SETD5", "ASH1L", "ARID1B", "DYRK1A"),
                            lfsr_mat = lfsr_mat1,
                            beta_mat = All_KOs.gibbs_PM$beta1_pm,
                            W_pm = All_KOs.gibbs_PM$W_pm)
```

## DEG GO enrichment
```{r}
enrich_folder <- paste0(wkdir, gsfa_folder, "WebGestalt_ORA/")
signif_term_clusters <- readRDS(paste0(enrich_folder, "go_bp.by_marker.signif_term_clusters.rds"))
signif_term_df <- signif_term_clusters$signif_term_df

abs_FC_colormap <- circlize::colorRamp2(breaks = c(0, 3, 6),
                                        colors = c("grey95", "#77d183", "#255566"))
```

### Early stage cells
```{r}
enrich_res <- readRDS(paste0(enrich_folder,
                             "All.gibbs_obj_k20.svd.go_bp.by_marker.group0.lfsr_0.05.rds"))
targets <- names(enrich_res)
enrich_table <- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) <- targets
for (i in 1:ncol(enrich_table)){
  m <- colnames(enrich_table)[i]
  enrich_df <- enrich_res[[m]]
  enrich_table[enrich_df$geneSet, i] <- enrich_df$enrichmentRatio
}
signif_term_df$degree <- rowSums(!is.na(enrich_table))

ordered_signif_term_df <- signif_term_df %>%
  filter(degree > 0, size >= 20) %>%
  arrange(group, -degree)
ordered_enrich_table <- enrich_table[ordered_signif_term_df$geneSet, ]
rownames(ordered_enrich_table) <- ordered_signif_term_df$description
```

```{r Early_GO_term_heatmap, fig.height=8, fig.width=8}
interest_enrich_table <- ordered_enrich_table
interest_enrich_table[is.na(interest_enrich_table)] <- 0

right_annot <- rowAnnotation(Cluster = ordered_signif_term_df$group,
                             col = list(Cluster = signif_term_clusters$colormap),
                             annotation_legend_param = list(
                               Cluster = list(
                                 title = "Cluster",
                                 at = signif_term_clusters$clustermap$group,
                                 labels = signif_term_clusters$clustermap$name
                               )
                             ),
                             show_legend = F)
map <- Heatmap(interest_enrich_table,
               name = "Fold of Enrichment",
               col = abs_FC_colormap,
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               right_annotation = right_annot,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               width = unit(8, "cm"))
draw(map, heatmap_legend_side = "right")
```

### Late stage cells
```{r}
enrich_res <- readRDS(paste0(enrich_folder,
                             "All.gibbs_obj_k20.svd.go_bp.by_marker.group1.lfsr_0.05.rds"))
targets <- names(enrich_res)
enrich_table <- data.frame(matrix(nrow = nrow(signif_term_df),
                                  ncol = length(targets)),
                           row.names = signif_term_df$geneSet)
colnames(enrich_table) <- targets
for (i in 1:ncol(enrich_table)){
  m <- colnames(enrich_table)[i]
  enrich_df <- enrich_res[[m]]
  enrich_table[enrich_df$geneSet, i] <- enrich_df$enrichmentRatio
}
signif_term_df$degree <- rowSums(!is.na(enrich_table))

ordered_signif_term_df <- signif_term_df %>%
  filter(degree > 0, size >= 20) %>%
  arrange(group, -degree)
ordered_enrich_table <- enrich_table[ordered_signif_term_df$geneSet, ]
rownames(ordered_enrich_table) <- ordered_signif_term_df$description
```

```{r Late_GO_term_heatmap, fig.height=13, fig.width=8}
interest_enrich_table <- ordered_enrich_table
interest_enrich_table[is.na(interest_enrich_table)] <- 0

right_annot <- rowAnnotation(Cluster = ordered_signif_term_df$group,
                             col = list(Cluster = signif_term_clusters$colormap),
                             annotation_legend_param = list(
                               Cluster = list(
                                 title = "Cluster",
                                 at = signif_term_clusters$clustermap$group,
                                 labels = signif_term_clusters$clustermap$name
                               )
                             ),
                             show_legend = F)
map <- Heatmap(interest_enrich_table,
               name = "Fold of Enrichment",
               col = abs_FC_colormap,
               row_title = NULL, column_title = NULL,
               cluster_rows = F, cluster_columns = F,
               show_row_dend = F, show_column_dend = F,
               right_annotation = right_annot,
               show_heatmap_legend = T,
               row_names_gp = gpar(fontsize = 10.5),
               column_names_rot = 45,
               width = unit(8, "cm"))
draw(map, heatmap_legend_side = "right")
```

# Gene Set Enrichment Analysis by Factor

Target genes: Genes w/ non-zero loadings in each factor (PIP cutoff at 0.95);    
Backgroud genes: all 6000 genes used in factor analysis;    
Statistical test: hypergeometric test (over-representation test);    
Only GO terms/pathways that satisfy fold change $\geq$ 2 and test FDR $<$ 0.05 are shown below.

```{r}
feature.names <- data.frame(fread(paste0(wkdir, "GSE142078_raw/GSM4219576_Run2_genes.tsv.gz"),
                                  header = FALSE), stringsAsFactors = FALSE)
names(feature.names) <- c("ID", "Name")
```

## GO Slim Over-Representation Analysis

Gene sets: [Gene ontology](http://www.geneontology.org/) "Biological Process" (non-redundant).

```{r go_bp}
gsfa.Web_ORA_go_bp <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.cutoff_0.95.rds"))
signif_num <- print_enrich_ORA_tb(gsfa.Web_ORA_go_bp, enrich_type = "GO terms",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Factors of interest

4, 9, 15

# Gene Set Enrichment Analysis by Perturbation

Target genes: Genes w/ LFSR < 0.05 under each perturbation marker gene;    
Backgroud genes: all 6000 genes used in factor analysis;    
Statistical test: hypergeometric test (over-representation test);    
Only GO terms/pathways that satisfy fold change $\geq$ 2 and test FDR $<$ 0.05 are shown below.

## GO Slim Over-Representation Analysis

Gene sets: [Gene ontology](http://www.geneontology.org/) "Biological Process" (non-redundant).

### Early stage cells
```{r go_bp_by_marker.group0}
Web_ORA_go_bp.by_marker <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.by_marker.group0.lfsr_0.05.rds"))
signif_num <- print_enrich_ORA_tb(Web_ORA_go_bp.by_marker,
                                  enrich_type = "GO terms", list_type = "per_marker",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Marker = names(Web_ORA_go_bp.by_marker),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(signif_num.df, caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Late stage cells
```{r go_bp_by_marker.group1}
Web_ORA_go_bp.by_marker <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "WebGestalt_ORA/All.gibbs_obj_k20.svd.go_bp.by_marker.group1.lfsr_0.05.rds"))
signif_num <- print_enrich_ORA_tb(Web_ORA_go_bp.by_marker,
                                  enrich_type = "GO terms", list_type = "per_marker",
                                  convert_genes = T, gene_map = feature.names)
signif_num.df <- t(data.frame(Marker = names(Web_ORA_go_bp.by_marker),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(signif_num.df, caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```
