---
title: "Guided Factor Analysis on TCGA BRCA Gene Expression Data"
subtitle: "-- Guides: Top 7 Somatic Mutation Genes"
author: "Yifan Zhou (zhouyf@uchicago.edu)"
date: '`r format(Sys.time(), "%Y-%m-%d")`'
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: true
---

```{r setup, include=FALSE}
requireNamespace("pander", quietly = TRUE)
library(data.table)
library(Matrix)
library(tidyverse)
library(ggplot2)
theme_set(theme_bw() + theme(plot.title = element_text(size = 14, hjust = 0.5),
                             axis.title = element_text(size = 14),
                             axis.text = element_text(size = 12),
                             legend.title = element_text(size = 13),
                             legend.text = element_text(size = 12),
                             panel.grid.minor = element_blank())
)
library(gridExtra)
library(ComplexHeatmap)
library(kableExtra)
library(clusterProfiler)

# set default chunk output
knitr::opts_chunk$set(echo = FALSE,
                      message = FALSE,
                      warning = FALSE,
                      comment = NA,
                      tidy = FALSE,
                      fig.width = 8,
                      fig.height = 6,
                      fig.align = "center",
                      results = "asis")

# formatting of pander tables
pander::panderOptions('knitr.auto.asis', FALSE)
pander::panderOptions("table.split.table", Inf)
```

```{r}
source("/project2/xinhe/yifan/Factor_analysis/reports/R/analytics_functions.R")
print_signif_tb <- function(signif_num, caption_text){
  signif_num <- t(data.frame(KO = names(signif_num),
                             Num_genes = signif_num,
                             row.names = NULL))
  print(knitr::kable(signif_num, caption = caption_text) %>%
    kable_styling() %>% scroll_box(width = '100%'))
}
print_lfsr_signif_tb <- function(lfsr_mat, lfsr_cutoff, caption_text){
  signif_num <- colSums(lfsr_mat < lfsr_cutoff)
  print_signif_tb(signif_num, caption_text)
}

wkdir <- "/project2/xinhe/yifan/TCGA/"
meta_file <- "processed_data/BRCA_metadata.female_white_only.rds"
gsfa_folder <- "gsfa_output_rsem_female_white/corrected/"
guide <- "All"
pip_col_fun <- circlize::colorRamp2(breaks = c(0, 0.5, 1),
                                    colors = c("black", "purple", "gold"))
```

# Data Description

References:
[Perspective on Oncogenic Processes at the End of the Beginning of Cancer Genomics](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5916814/).

Data source:   
[FireBrowse TCGA BRCA Archives](http://firebrowse.org/?cohort=BRCA&download_dialog=true).

Perturbations:    
Somatic mutation status of selected frequently mutated driver genes (PIK3CA, TP53, TTN, GATA3, CDH1, MAP3K1, MAP2K4).

Samples:    
TCGA breast invasive carcinoma (BRCA) tumor samples.   
To avoid possible confounding due to gender and race, **we confined our study samples to be only from female Caucasian subjects**.    
In addition, only samples with somatic mutation annotation were kept, resulted in 700 samples.

Genes:    
**Only the top 10k variable genes in samples of consideration were kept.**

Normalization:    
**RSEM normalized counts corrected for subject age and tumor purity.** The corrected and scaled expression data were used as input for subsequent factor analysis.

```{r load meta data, fig.width=9}
metadata <- readRDS(paste0(wkdir, meta_file))
all_KOs.G_mat <- metadata[, 14:20]
KO_names <- colnames(all_KOs.G_mat)
nmarkers <- ncol(all_KOs.G_mat)
```

# Guided results

Here, our "guide", $G$ matrix, consists of the somatic mutation status of top 7 driver genes across samples.

We specified 20 factors in our model. In each case, Gibbs sampling was conducted for 2000 iterations, and the posterior mean estimates were averaged over the last 500 iterations.

## SVD Initialization
```{r svd_tb}
All_KOs.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.PM.rds"))
stopifnot(nrow(All_KOs.gibbs_PM$beta_pm) == nmarkers + 1)
All_KOs.gibbs_res_tb <- make_gibbs_res_tb(All_KOs.gibbs_PM, all_KOs.G_mat, compute_pve = F)
```

### Local False Sign Rate (LFSR)

For a given GSFA inference result, we can estimate the effect a certain KO condition $m$ has on the expression of gene $j$ by computing the LFSR of $\beta_{m\cdot} \cdot W_{j\cdot}$.

```{r}
lfsr_mat <- readRDS(paste0(wkdir, gsfa_folder, guide, ".gibbs_obj_k20.svd.lfsr_mat.rds"))
print_lfsr_signif_tb(lfsr_mat, 0.05,
                     paste0("GSFA, # of genes that passed LFSR cutoff of 0.05:"))
```

### Factor ~ KO Beta PIP
```{r fig.width=8, fig.height=6}
beta_pip_matrix <- t(All_KOs.gibbs_PM$Gamma_pm[-(nmarkers + 1), ])
rownames(beta_pip_matrix) <- 1:nrow(beta_pip_matrix)
colnames(beta_pip_matrix) <- colnames(all_KOs.G_mat)
Heatmap(beta_pip_matrix, name = "Beta PIP", col = pip_col_fun,
        row_title = "GSFA Factors",  column_title = "Somatic Mutations",
        cluster_rows = F, cluster_columns = F)
```

### Factor ~ KO Posterior Association
```{r fig.width=9, fig.height=6}
heatmap_matrix <- All_KOs.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = All_KOs.gibbs_res_tb$pi,
                  row_title = "GSFA Factors", column_title = "Somatic Mutations")
guided.heatmap_matrix <- heatmap_matrix
gsfa_KO_names <- apply(guided.heatmap_matrix, 1,
                       function(x){ paste(colnames(all_KOs.G_mat)[x < 1e-4], collapse = ",") })
gsfa_KO_names <- paste(1:nrow(guided.heatmap_matrix), gsfa_KO_names, sep = " - ")
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA All Mutations (K=20, SVD)")
```

### Beta PIP vs P-Value
```{r fig.width=6, fig.height=5}
qplot(unlist(beta_pip_matrix), -log10(unlist(heatmap_matrix)),
            xlab = "Beta PIP", ylab = "-log10(Association p-value)",
            main = "20 factors x 7 conditions\n(red line at y=5, omiting 1 data point at y=68)") +
  geom_hline(yintercept = 5, color = "red", linetype = "dashed") +
  scale_y_continuous(limits = c(0, 20))
```

### Correlation btw Factors

```{r eval=FALSE, include=FALSE}
plot_pairwise.corr_heatmap(input_mat_1 = All_KOs.gibbs_PM$Z_pm,
                           corr_type = "pearson",
                           name_1 = "Pairwise Correlation in Sample Loadings - Z (GSFA)")
```

```{r}
plot_pairwise.corr_heatmap(input_mat_1 = (All_KOs.gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Binarized Gene Loadings (F.pm > 0.95)")
```

# Gene Ontology Enrichment Analysis

Statistical test: hypergeometric test;

GO category: Biological Process.

## GO enrichment per factor

Target: Genes w/ PIP > 0.95 in each factor;    
Backgroud: all 10000 genes used in factor analysis.

GO terms that passed over-representation test fold change $\geq$ 1.5 and q value $<$ 0.05:

```{r}
gsfa.go_enrich <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "enrich_res/All.gibbs_obj_k20.svd.go_enrich.cutoff_0.95.rds"))
signif_num <- print_enrich_tb(gsfa.go_enrich, qvalue_cutoff = 0.05, FC_cutoff = 1.5,
                              type = "per_factor")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

### Factors of interest

8: immune response    
11: translation   
13: blood vessel development; epidermal cell differentiation    
16: RNA splicing    
20: cell cycle; nuclear division

## GO enrichment per driver gene condition

Target: Genes w/ GSFA lfsr < 0.05 under each condition, stratefied by sign of effect;    
Backgroud: all 10000 genes used in factor analysis.

GO terms that passed over-representation test fold change $\geq$ 1.5 and q value $<$ 0.05:

```{r}
gsfa.go_enrich_permarker <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "enrich_res/All.gibbs_obj_k20.svd.go_enrich.permarker_updown.rds"))
signif_num <- print_enrich_tb(gsfa.go_enrich_permarker, qvalue_cutoff = 0.05, FC_cutoff = 1.5,
                              type = "per_marker")
signif_num.df <- t(data.frame(Mutation_status = names(gsfa.go_enrich_permarker),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(signif_num.df, caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

# Unguided Result

In contrast to the guided version, here we provide no guidance ($G = \vec{0}$) to our sparse factor analysis model to serve as a comparison. 
```{r G0_tb}
G0.gibbs_PM <- readRDS(paste0(wkdir, gsfa_folder, "G0.gibbs_obj_k20.svd.PM.rds"))
G0.gibbs_res_tb <- make_gibbs_res_tb(G0.gibbs_PM, all_KOs.G_mat, compute_pve = F)
```

## Factor ~ KO Beta PIP
```{r fig.width=8, fig.height=1}
beta_pip_matrix <- t(G0.gibbs_PM$Gamma_pm[-2, ])
colnames(beta_pip_matrix) <- 1:ncol(beta_pip_matrix)
rownames(beta_pip_matrix) <- "G0"
Heatmap(beta_pip_matrix,
        name = "Beta PIP", col = pip_col_fun,column_title = "Factors",
        cluster_rows = F, cluster_columns = F,
        heatmap_legend_param = list(legend_direction = "horizontal"))
```

## Factor ~ KO Posterior Association
```{r fig.width=10, fig.height=6}
heatmap_matrix <- G0.gibbs_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = G0.gibbs_res_tb$pi,
                  row_title = "GSFA Factors", column_title = "Somatic Mutations")
G0.heatmap_matrix <- heatmap_matrix
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "GSFA Unguide All KOs (K=20, SVD)")
```

## Unguided vs Guided

### Association p value comparison

```{r fig.width=10}
paired_pval_ranked_scatterplot(pval_vec_1 = unlist(guided.heatmap_matrix),
                               pval_vec_2 = unlist(G0.heatmap_matrix),
                               name_1 = "Guided", name_2 = "Unguided", zoom_in_y = 10)
```

### Correlation btw guided & unguided factors

```{r}
plot_pairwise.corr_heatmap(input_mat_1 = (All_KOs.gibbs_PM$F_pm > 0.95) * 1,
                           input_mat_2 = (G0.gibbs_PM$F_pm > 0.95) * 1,
                           corr_type = "jaccard",
                           name_1 = "Binarized Gene Loadings (Guided)",
                           name_2 = "Binarized Gene Loadings (Unguided)")
```

## GO enrichment per factor (unguided)

Target: Genes w/ PIP > 0.95 in each factor;    
Backgroud: all 10000 genes used in factor analysis.

GO terms that passed over-representation test fold change $\geq$ 1.5 and q value $<$ 0.05:

```{r}
gsfa.go_enrich <-
  readRDS(paste0(wkdir, gsfa_folder,
                 "enrich_res/G0.gibbs_obj_k20.svd.go_enrich.cutoff_0.95.rds"))
signif_num <- print_enrich_tb(gsfa.go_enrich, qvalue_cutoff = 0.05, FC_cutoff = 1.5,
                              type = "per_factor")
signif_num.df <- t(data.frame(Factor = 1:length(signif_num),
                              Signif_GO_terms = signif_num,
                              row.names = NULL))
knitr::kable(rbind(signif_num.df[, 1:10], signif_num.df[, 11:20]),
             caption = "Summary:") %>%
  kable_styling() %>% scroll_box(width = '100%')
```

# FLASH

```{r flash_tb}
flash_obj <-
  readRDS(paste0(wkdir, "flash_output/BRCA_RSEM_w_mutation.female_white_only.top10k_var_scaled.corrected.flashier_obj_k20.rds"))
flash_res_tb <- make_flash_res_tb(flash_obj, all_KOs.G_mat)
```

## Factor ~ KO Posterior Association

```{r FLASH pval heatmap, fig.width=10}
heatmap_matrix <- flash_res_tb %>% select(starts_with("pval"))
rownames(heatmap_matrix) <- 1:nrow(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(all_KOs.G_mat)
plot_pval_heatmap(heatmap_matrix,
                  factor_annot = flash_res_tb$pi1,
                  row_title = "FLASH Factors", column_title = "Somatic Mutations")
flash_heatmap <- heatmap_matrix
```

```{r fig.width=10, fig.height=5}
summ_pvalues(unlist(heatmap_matrix),
             title_text = "FLASH All KOs (K=20, SVD)")
```

## FLASH vs GSFA

```{r fig.width=8}
paired_pval_ranked_scatterplot(pval_vec_1 = unlist(guided.heatmap_matrix),
                               pval_vec_2 = unlist(flash_heatmap),
                               name_1 = "Guided", name_2 = "FLASH")
```

We matched the factors obtained from GFSA with those from FLASH, and conducted pairwise comparisons for both factors and gene loadings.

**In the following heatmaps, factors are labeled by the KO condition(s) that they have an association p-value < 1e-4 with.**

```{r}
flash_KO_names <- apply(flash_heatmap, 1,
                        function(x){ paste(colnames(all_KOs.G_mat)[x < 1e-4], collapse = ",") })
flash_KO_names <- paste(1:nrow(flash_heatmap), flash_KO_names, sep = " - ")
# flash_matching_order <- apply(corr_mat, 1, function(x){ which.max(abs(x)) })
flash_matching_order <- c(1:13, 15, 14, 15, 17, 18, 20, 2)
```

```{r fig.width=9, fig.height=7}
Z_pm.flash <- flash_obj$loadings.pm[[1]][, flash_matching_order]
colnames(Z_pm.flash) <- flash_KO_names[flash_matching_order]
Z_pm.gsfa <- All_KOs.gibbs_PM$Z_pm
colnames(Z_pm.gsfa) <- gsfa_KO_names
plot_pairwise.corr_heatmap(input_mat_1 = Z_pm.gsfa,
                           input_mat_2 = Z_pm.flash,
                           name_1 = "GSFA Factors (Z_pm)", name_2 = "FLASH Factors")
```

```{r fig.width=9, fig.height=7}
F_pm.flash <- flash_obj$loadings.lfsr[[2]][, flash_matching_order]
colnames(F_pm.flash) <- flash_KO_names[flash_matching_order]
F_pm.gsfa <- All_KOs.gibbs_PM$F_pm
colnames(F_pm.gsfa) <- gsfa_KO_names
plot_pairwise.corr_heatmap(input_mat_1 = F_pm.gsfa > 0.95,
                           input_mat_2 = F_pm.flash < 0.05,
                           name_1 = "GSFA Binarized Gene Loadings (F_pm > 0.95)",
                           name_2 = "FLASH Binarized Gene Loadings (LFSR < 0.05)",
                           corr_type = "jaccard")
```
