<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Yifan Zhou (zhouyf@uchicago.edu)" />

<meta name="date" content="2020-11-12" />

<title>Factor Analysis on MCF10A CROP-seq Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}

.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Intro</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Factor Analysis on MCF10A CROP-seq Data</h1>
<h4 class="author"><em>Yifan Zhou (<a href="mailto:zhouyf@uchicago.edu">zhouyf@uchicago.edu</a>)</em></h4>
<h4 class="date"><em>2020-11-12</em></h4>

</div>


<div id="data-settings" class="section level1">
<h1><span class="header-section-number">1</span> Data Settings</h1>
<p>CROP-seq data source:<br />
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5882576/">On the design of CRISPR-based single cell molecular screens</a>, GEO accession: GSE108699.</p>
<p>Cells:<br />
MCF10A cells (normal human breast epithelial cells) <strong>with exposure to a DNA damaging agent, doxorubicin</strong>;</p>
<p>Perturbations:<br />
CRISPR knock-outs of 29 tumor-suppressor genes (TP53, …), 1 non-targeting control.</p>
<div id="doxorubicin-treated-expression-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Doxorubicin-treated expression data</h2>
<p>There are 5584 cells with nonzero gRNA count.</p>
</div>
<div id="perturbations" class="section level2">
<h2><span class="header-section-number">1.2</span> Perturbations</h2>
<p><img src="index_files/figure-html/load%20meta%20data-1.png" width="864" style="display: block; margin: auto;" /><img src="index_files/figure-html/load%20meta%20data-2.png" width="864" style="display: block; margin: auto;" /></p>
<div id="distribution-of-knock-out-perturbations" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Distribution of Knock-out Perturbations</h3>
<p><img src="figs/umap_KO_patterns.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="wilcoxon-differential-expression-test" class="section level3">
<h3><span class="header-section-number">1.2.2</span> Wilcoxon Differential Expression Test</h3>
Number of genes that passed pval_adjust &lt; 0.2 under each perturbation:
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<tbody>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
APC
</td>
<td style="text-align:left;">
ARID1A
</td>
<td style="text-align:left;">
ARID1B
</td>
<td style="text-align:left;">
ATM
</td>
<td style="text-align:left;">
BAP1
</td>
<td style="text-align:left;">
BID
</td>
<td style="text-align:left;">
BRCA1
</td>
<td style="text-align:left;">
BRCA2
</td>
<td style="text-align:left;">
CASP8
</td>
<td style="text-align:left;">
CDH1
</td>
</tr>
<tr>
<td style="text-align:left;">
DE_genes
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
CDKN1B
</td>
<td style="text-align:left;">
CDKN2A
</td>
<td style="text-align:left;">
CHEK1
</td>
<td style="text-align:left;">
CTCF
</td>
<td style="text-align:left;">
GATA3
</td>
<td style="text-align:left;">
MLH1
</td>
<td style="text-align:left;">
MLL2
</td>
<td style="text-align:left;">
MLL3
</td>
<td style="text-align:left;">
NCOR1
</td>
<td style="text-align:left;">
NF1
</td>
</tr>
<tr>
<td style="text-align:left;">
DE_genes
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KO
</td>
<td style="text-align:left;">
NONTARGETING
</td>
<td style="text-align:left;">
PTEN
</td>
<td style="text-align:left;">
PTPN22
</td>
<td style="text-align:left;">
PTPRD
</td>
<td style="text-align:left;">
RAD51
</td>
<td style="text-align:left;">
RB1
</td>
<td style="text-align:left;">
RUNX1
</td>
<td style="text-align:left;">
SETD2
</td>
<td style="text-align:left;">
STK11
</td>
<td style="text-align:left;">
TP53
</td>
</tr>
<tr>
<td style="text-align:left;">
DE_genes
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
33
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
34
</td>
<td style="text-align:left;">
3476
</td>
</tr>
</tbody>
</table>
</div>
<p><img src="index_files/figure-html/pval_FC_plot-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="potential-confounding-factors" class="section level2">
<h2><span class="header-section-number">1.3</span> Potential confounding factors</h2>
<p>Positive correlation between number of guides and cell library size: <img src="index_files/figure-html/libsize_confound-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Cell cycle effects (colored by the mean expression of signature genes at the corresponding cell cycle stage):</p>
<p><img src="figs/umap_cell_cycle_patterns.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="preprocessing-of-rna-count-matrix" class="section level2">
<h2><span class="header-section-number">1.4</span> Preprocessing of RNA count matrix:</h2>
<p>Seurat “LogNormalize”: log(count per <span class="math inline">\(10^4\)</span> + 1).</p>
<p><em>Library size was regressed out, and the residuals were used as input.</em></p>
<p>Only cells with gRNA readout were kept.</p>
<p>We tried two ways to filter the genes.</p>
</div>
</div>
<div id="factor-analysis-on-expression-data" class="section level1">
<h1><span class="header-section-number">2</span> Factor Analysis on Expression Data</h1>
<div id="using-top-10k-most-variable-genes" class="section level2">
<h2><span class="header-section-number">2.1</span> Using top 10K most variable genes</h2>
<p><a href="gsfa_all_markers_top10k_var.html">GSFA all markers result</a>;</p>
<p><a href="unguided_top10k_var.html">Unguided G=0 result</a>;</p>
<p><a href="flash_top10k_var.html">FLASH result</a>.</p>
</div>
<div id="using-top-8k-most-variable-genes" class="section level2">
<h2><span class="header-section-number">2.2</span> Using top 8K most variable genes</h2>
<p><a href="gsfa_all_markers_top8k_var.html">GSFA all markers result</a>;</p>
<p>Unguided G=0 result;</p>
<p><a href="flash_top8k_var.html">FLASH result</a>.</p>
</div>
<div id="using-genes-detected-in-5-of-cells" class="section level2">
<h2><span class="header-section-number">2.3</span> Using genes detected in &gt; 5% of cells</h2>
<p><a href="gsfa_all_markers_detect_005.html">GSFA all markers result</a>;</p>
<p><a href="unguided_detect_005.html">Unguided G=0 result</a>;</p>
<p><a href="flash_detect_005.html">FLASH result</a>.</p>
</div>
<div id="using-genes-detected-in-10-of-cells" class="section level2">
<h2><span class="header-section-number">2.4</span> Using genes detected in &gt; 10% of cells</h2>
<p><a href="gsfa_all_markers_detect_01.html">GSFA all markers result</a>;</p>
<p><a href="unguided_detect_01.html">Unguided G=0 result</a>;</p>
<p><a href="flash_detect_01.html">FLASH result</a>.</p>
</div>
<div id="permutation-test-on-associations" class="section level2">
<h2><span class="header-section-number">2.5</span> Permutation test on associations</h2>
<p><a href="permutation.html">Permutation results for all methods</a>.</p>
</div>
<div id="gene-discovery-comparison-using-local-false-sign-rate" class="section level2">
<h2><span class="header-section-number">2.6</span> Gene discovery comparison using local false sign rate</h2>
<p>For each factor analysis result, we estimated the KO effect on genes using the local false sign rate (LFSR), and compared the result with that from Wilcoxon DGE.</p>
<p><a href="lfsr.html">LFSR result comparison</a>.</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
