<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yifan Zhou (zhouyf@uchicago.edu)" />

<meta name="date" content="2021-04-04" />

<title>Functional Analysis on LUHMES CROP-seq GSFA Results</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSFA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Functional Analysis on LUHMES CROP-seq GSFA Results</h1>
<h3 class="subtitle">-- NTC included, unique UMI count corrected</h3>
<h4 class="author">Yifan Zhou (<a href="mailto:zhouyf@uchicago.edu">zhouyf@uchicago.edu</a>)</h4>
<h4 class="date">2021-04-04</h4>

</div>


<div id="gsfa-on-luhmes-crop-seq-data" class="section level1">
<h1><span class="header-section-number">1</span> GSFA on LUHMES CROP-seq Data</h1>
<div id="data-processing" class="section level2">
<h2><span class="header-section-number">1.1</span> Data Processing</h2>
<p>Cells:<br />
LUHMES cells from 3 batches were merged together into 1 analysis. All cells have only a single type of gRNA readout. Quality control resulted in 8708 cells.</p>
<p>Genes:<br />
<strong>Only genes detected in &gt; 10% of cells were kept, resulted in 6213 genes.</strong></p>
<p>Normalization:<br />
Seurat &quot;LogNormalize&quot;: log(count per 10K + 1).<br />
<strong>Batch effect, unique UMI count, library size, and mitochondria percentage were all corrected for. The corrected and scaled expression data were used as input for subsequent factor analysis.</strong></p>
</div>
<div id="gsfa-results-svd-initialization" class="section level2">
<h2><span class="header-section-number">1.2</span> GSFA Results (SVD Initialization)</h2>
<p>Here, our &quot;guide&quot;, <span class="math inline">\(G\)</span> matrix, consists of 15 types (14 genes + NTC) of gene-level knock-down conditions across cells.</p>
<p>Gibbs sampling was initiated from SVD, and conducted for 2000 iterations; the posterior mean estimates were averaged over the last 500 iterations.</p>
<div id="estimate-of-factor-perturbation-associations-beta" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Estimate of Factor ~ Perturbation Associations (<span class="math inline">\(\beta\)</span>)</h3>
<p><img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/unnamed-chunk-7-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Examples of factor ~ perturbation associations: <img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="degs-lfsr-0.05-under-each-perturbation" class="section level3">
<h3><span class="header-section-number">1.2.2</span> DEGs (LFSR &lt; 0.05) under Each Perturbation</h3>
<p>Number of genes that passed GSFA LFSR &lt; 0.05 under each perturbation:</p>
<p>Number of genes that passed t-test FDR &lt; 0.05 under each perturbation:</p>
<p><img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/dge_comparison_log-1.png" width="960" style="display: block; margin: auto;" /></p>
<p><img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/dge_comparison-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Overlap of GSFA DEGs between perturbations:</p>
<p><img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/unnamed-chunk-10-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Venn diagram for DEGs found under perturbations ADNP, ARID1B, ASH1L, CHD2:</p>
<p><img src="figs/LUHMES_DEGs_venn_diagram_4.png" width="60%" style="display: block; margin: auto;" /></p>
<p>SFARI high risk (score 1, 2) genes among the DEGs: <img src="figs/LUHMES_Marker_SFARI_DEG_network.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="gene-set-enrichment-analysis" class="section level1">
<h1><span class="header-section-number">2</span> Gene Set Enrichment Analysis</h1>
<div id="enrichment-by-factor" class="section level2">
<h2><span class="header-section-number">2.1</span> Enrichment by factor</h2>
<p>Target genes: Genes w/ non-zero loadings in each factor (PIP cutoff at 0.95);<br />
Backgroud genes: all 6213 genes included in factor analysis;<br />
Statistical test: hypergeometric test (over-representation test);<br />
Only GO terms/pathways within the size of 10 ~ 500 and have have an enrichment FC &gt; 2 and FDR &lt; 0.05 are kept.</p>
<div id="go-slim-over-representation-analysis" class="section level3">
<h3><span class="header-section-number">2.1.1</span> GO Slim Over-Representation Analysis</h3>
<p>Gene sets: <a href="http://www.geneontology.org/">Gene ontology</a> &quot;Biological Process&quot; (non-redundant).</p>
<p>We used the &quot;Wang&quot; method in <code>GOSemSim</code> to measure the similarity between GO BP terms, and all the significant terms in factors of interest were further grouped into 9 clusters using hierarchical clustering with the &quot;ward.D&quot; agglomeration method. The clustering results of all these GO BP terms are stored <a href="https://drive.google.com/drive/folders/1yady1-Z_ehJXkfqZxe_CtqEZdFN1TD7Z">here</a>.</p>
<p><strong>Terms of interest</strong></p>
<p><img src="figs/LUHMES_foldchange_heatmap_interest_terms_annotated.png" width="100%" style="display: block; margin: auto;" /></p>
<p><strong>Factors of interest</strong></p>
<ul>
<li><a href="http://www.webgestalt.org/results/1614731260/#">Factor 10</a>: nerve development, cell fate commitment;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614731201/#">Factor 11</a>: chromatin remodeling, DNA recombination;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614731171/#"><strong>Factor 13</strong></a>: regulation of metal ion transport, axon development, ...;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614731117/#"><strong>Factor 15</strong></a>: neuron migration, response to metal ion, CNS neuron development, ...;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614730861/#"><strong>Factor 18</strong></a>: CNS neuron differentiation, xxx tissue development, ...;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614730760/#">Factor 19</a>: axon development;<br />
</li>
<li><a href="http://www.webgestalt.org/results/1614730397/#"><strong>Factor 20</strong></a>: nerve development, neurotransmitter transport, ....</li>
</ul>
</div>
<div id="reactome-pathway-over-representation-analysis" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Reactome Pathway Over-Representation Analysis</h3>
<p>Gene sets: The <a href="http://www.reactome.org/">Reactome pathway database</a>.</p>
<p><strong>Factors of interest</strong></p>
<ul>
<li><a href="http://www.webgestalt.org/results/1614728526/#">Factor 8</a>: p53-Independent DNA Damage Response, G2/M Transition</li>
<li><a href="http://www.webgestalt.org/results/1614728715/#">Factor 9</a>: Neuronal System, Transmission across Chemical Synapses</li>
<li><a href="http://www.webgestalt.org/results/1614728793/#">Factor 12</a>: Gap junction trafficking</li>
<li><a href="http://www.webgestalt.org/results/1614728429/#">Factor 13</a>: Neurotransmitter receptors and postsynaptic signal transmission, Ion channel transport, Neuronal System</li>
<li><a href="http://www.webgestalt.org/results/1614728922/#">Factor 15</a>: Transmission across Chemical Synapses, Ca2+ pathway, Gap junction trafficking</li>
<li><a href="http://www.webgestalt.org/results/1614729171/#">Factor 17</a>: Axon guidance</li>
<li><a href="http://www.webgestalt.org/results/1614729226/#">Factor 19</a>: Activation of NMDA receptors and postsynaptic events, Mitotic phases, Cell Cycle Checkpoints</li>
<li><a href="http://www.webgestalt.org/results/1614729303/#">Factor 20</a>: Neurotransmitter receptors and postsynaptic signal transmission, Ion channel transport</li>
</ul>
</div>
</div>
</div>
<div id="inspection-of-signature-genes" class="section level1">
<h1><span class="header-section-number">3</span> Inspection of Signature Genes</h1>
<div id="genes-targeted-by-crispr-knock-down" class="section level2">
<h2><span class="header-section-number">3.1</span> Genes targeted by CRISPR knock-down</h2>
<p>PTEN is not detected in &gt; 10% cells.</p>
</div>
<div id="neuron-projection-development-genes" class="section level2">
<h2><span class="header-section-number">3.2</span> Neuron projection development genes</h2>
<p>According to Figure 4E of the <a href="https://genome.cshlp.org/content/30/9/1317.full">reference paper</a>, among these marker genes, mature neuron marker genes are down-regulated and negative regulator genes are up-regulated in ADNP, CHD2 and ASH1L knock-downs; while the opposite happens in PTEN knock-down.</p>
<p>Here we show all the marker genes that passed LFSR &lt; 0.05 under each knock-down. In general, their effect sizes are consistent with the findings of the paper.</p>
<p><img src="LUHMES_merged_new.gsfa_all_markers_detect_01_enrichment_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
